

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Text Outputs &mdash; Tombo 1.2 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="index" title="Index"
              href="genindex.html"/>
        <link rel="search" title="Search" href="search.html"/>
    <link rel="top" title="Tombo 1.2 documentation" href="index.html"/>
        <link rel="next" title="Plotting Commands" href="plotting.html"/>
        <link rel="prev" title="Modified Base Detection" href="modified_base_detection.html"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> Tombo
          

          
          </a>

          
            
            
              <div class="version">
                1.2
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="examples.html">Tombo Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="resquiggle.html">Re-squiggle Algorithm</a></li>
<li class="toctree-l1"><a class="reference internal" href="modified_base_detection.html">Modified Base Detection</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Text Outputs</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#write-wiggles"><code class="docutils literal notranslate"><span class="pre">write_wiggles</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#write-most-significant-fasta"><code class="docutils literal notranslate"><span class="pre">write_most_significant_fasta</span></code></a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="plotting.html">Plotting Commands</a></li>
<li class="toctree-l1"><a class="reference internal" href="filtering.html">Read Filtering Commands</a></li>
<li class="toctree-l1"><a class="reference internal" href="model_training.html">Model Training (Advanced Users Only)</a></li>
<li class="toctree-l1"><a class="reference internal" href="rna.html">RNA Processing</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="tombo.html">tombo package</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Tombo</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Text Outputs</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/text_output.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="text-outputs">
<h1>Text Outputs<a class="headerlink" href="#text-outputs" title="Permalink to this headline">¶</a></h1>
<p>Two text outputs are available from Tombo:</p>
<ol class="arabic simple">
<li>Wiggle - Genome browser compatible run per-base statistics</li>
<li>Fasta - Sequence output surrounding most modified sites</li>
</ol>
<div class="section" id="write-wiggles">
<h2><code class="docutils literal notranslate"><span class="pre">write_wiggles</span></code><a class="headerlink" href="#write-wiggles" title="Permalink to this headline">¶</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">write_wiggles</span></code> command takes in a set of reads (<code class="docutils literal notranslate"><span class="pre">--fast5-basedirs</span></code>) and/or a pre-computed statistics file (<code class="docutils literal notranslate"><span class="pre">--statistics-filename</span></code>). A control set of reads can also be provided (<code class="docutils literal notranslate"><span class="pre">--control-fast5-basedirs</span></code>). Output wiggle files (<a class="reference external" href="https://genome.ucsc.edu/goldenpath/help/wiggle.html">variableStep format</a>) will be produced for each requested statistic (both plus and minus strands).</p>
<p>Several statistics are available for output:</p>
<ul class="simple">
<li><code class="docutils literal notranslate"><span class="pre">coverage</span></code> - The coverage level for mapped and validly re-squiggled reads</li>
<li><code class="docutils literal notranslate"><span class="pre">dampened_fraction</span></code> - The estimated fraction of significantly modified reads<ul>
<li>This estimate includes pseudo-counts added to the un-modified and modified read counts (as specified by the <code class="docutils literal notranslate"><span class="pre">--coverage-dampen-counts</span></code> option)</li>
<li>This is equivalent to using a beta prior when estimating the fraction of reads modified at this position</li>
<li>Test the effect of different dampen counts using the <code class="docutils literal notranslate"><span class="pre">scripts/test_beta_priors.R</span></code> (the default values are shown below)</li>
</ul>
</li>
<li><code class="docutils literal notranslate"><span class="pre">fraction</span></code> - The raw fraction of significantly modified reads</li>
<li><code class="docutils literal notranslate"><span class="pre">signal</span></code> - The mean signal level across all reads mapped to this location</li>
<li><code class="docutils literal notranslate"><span class="pre">signal_sd</span></code> - The mean signal standard deviation across all reads mapped to this location (not available unless <code class="docutils literal notranslate"><span class="pre">--include-event-stdev</span></code> was provided in resquiggle call)</li>
<li><code class="docutils literal notranslate"><span class="pre">dwell</span></code> - The mean number of raw observations observed assigned to this location</li>
<li><code class="docutils literal notranslate"><span class="pre">difference</span></code> - The difference in normalized signal level between a sample and control set of reads</li>
</ul>
<hr class="docutils" />
<div class="figure align-center" id="id1">
<a class="reference internal image-reference" href="_images/dampened_fraction.png"><img alt="_images/dampened_fraction.png" src="_images/dampened_fraction.png" style="width: 627.6px; height: 546.0px;" /></a>
<p class="caption"><span class="caption-text">Heatmap showing the resulting dampened farction of modified reads given the default <code class="docutils literal notranslate"><span class="pre">--coverage-dampen-counts</span></code> values over range of coverage and number of un-modified reads.</span></p>
</div>
<hr class="docutils" />
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last"><code class="docutils literal notranslate"><span class="pre">signal</span></code>, <code class="docutils literal notranslate"><span class="pre">signal_sd</span></code>, <code class="docutils literal notranslate"><span class="pre">dwell</span></code> and <code class="docutils literal notranslate"><span class="pre">difference</span></code> require each reads event level data to be queried and thus may be quite slow. <code class="docutils literal notranslate"><span class="pre">coverage</span></code>, <code class="docutils literal notranslate"><span class="pre">dampened_fraction</span></code>, and <code class="docutils literal notranslate"><span class="pre">fraction</span></code> can be extracted simply from the tombo statistics file, which is much faster.</p>
</div>
<p>Files will be output to individual wiggle files (two per statistic for plus and minus genomic strand) in the following format <code class="docutils literal notranslate"><span class="pre">[wiggle-basename].[wiggle-type].[sample|control]?.[plus|minus].wig</span></code></p>
</div>
<div class="section" id="write-most-significant-fasta">
<h2><code class="docutils literal notranslate"><span class="pre">write_most_significant_fasta</span></code><a class="headerlink" href="#write-most-significant-fasta" title="Permalink to this headline">¶</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">write_most_significant_fasta</span></code> command writes the genome sequence surrounding the most modified positions. This can be useful for several tasks related to modified base detection including motif discovery.</p>
<p>To run <code class="docutils literal notranslate"><span class="pre">write_most_significant_fasta</span></code>, a <code class="docutils literal notranslate"><span class="pre">--statistics-filename</span></code> is required to extract the most significant locations and either a <code class="docutils literal notranslate"><span class="pre">--fast5-basedirs</span></code> or <code class="docutils literal notranslate"><span class="pre">--genome-fasta</span></code> is required to extract the genomic sequence. Several options are availble for selecting the sequence to be output:</p>
<ul class="simple">
<li><code class="docutils literal notranslate"><span class="pre">--num-regions</span></code> - Defines the number of unique locations to be output</li>
<li><code class="docutils literal notranslate"><span class="pre">--num-bases</span></code> - Defines the number of bases to be output surrounding the significant locations</li>
</ul>
<p>The output of this command could be used to determine sequence contexts consistently modified within a sample. Example <a class="reference external" href="http://meme-suite.org/doc/meme.html">meme</a> command line modified base motif detection command.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>./meme -oc motif_output.meme -dna -mod zoops tombo_results.significant_regions.fasta
</pre></div>
</div>
</div>
</div>


           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="plotting.html" class="btn btn-neutral float-right" title="Plotting Commands" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="modified_base_detection.html" class="btn btn-neutral" title="Modified Base Detection" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, Oxford Nanopore Technologies.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'1.2',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>