

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>tombo.tombo_helper &mdash; Tombo 1.2 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="index" title="Index"
              href="../../genindex.html"/>
        <link rel="search" title="Search" href="../../search.html"/>
    <link rel="top" title="Tombo 1.2 documentation" href="../../index.html"/>
        <link rel="up" title="Module code" href="../index.html"/> 

  
  <script src="../../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../../index.html" class="icon icon-home"> Tombo
          

          
          </a>

          
            
            
              <div class="version">
                1.2
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../examples.html">Tombo Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../resquiggle.html">Re-squiggle Algorithm</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../modified_base_detection.html">Modified Base Detection</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../text_output.html">Text Outputs</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../plotting.html">Plotting Commands</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../filtering.html">Read Filtering Commands</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../model_training.html">Model Training (Advanced Users Only)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../rna.html">RNA Processing</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../tombo.html">tombo package</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">Tombo</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../index.html">Module code</a> &raquo;</li>
        
      <li>tombo.tombo_helper</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for tombo.tombo_helper</h1><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">__future__</span> <span class="k">import</span> <span class="n">division</span><span class="p">,</span> <span class="n">unicode_literals</span><span class="p">,</span> <span class="n">absolute_import</span>

<span class="kn">from</span> <span class="nn">builtins</span> <span class="k">import</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">range</span><span class="p">,</span> <span class="nb">dict</span><span class="p">,</span> <span class="nb">map</span><span class="p">,</span> <span class="nb">zip</span>

<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">io</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">random</span>
<span class="kn">import</span> <span class="nn">fnmatch</span>

<span class="c1"># Future warning from cython in h5py</span>
<span class="kn">import</span> <span class="nn">warnings</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">simplefilter</span><span class="p">(</span><span class="n">action</span><span class="o">=</span><span class="s1">&#39;ignore&#39;</span><span class="p">,</span> <span class="n">category</span><span class="o">=</span><span class="ne">FutureWarning</span><span class="p">)</span>
<span class="kn">import</span> <span class="nn">h5py</span>

<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="kn">from</span> <span class="nn">glob</span> <span class="k">import</span> <span class="n">glob</span>
<span class="kn">from</span> <span class="nn">operator</span> <span class="k">import</span> <span class="n">itemgetter</span>
<span class="kn">from</span> <span class="nn">itertools</span> <span class="k">import</span> <span class="n">repeat</span><span class="p">,</span> <span class="n">islice</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="k">import</span> <span class="n">defaultdict</span><span class="p">,</span> <span class="n">namedtuple</span>

<span class="k">if</span> <span class="n">sys</span><span class="o">.</span><span class="n">version_info</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">2</span><span class="p">:</span>
    <span class="n">unicode</span> <span class="o">=</span> <span class="nb">str</span>

<span class="c1"># import tombo functions</span>
<span class="kn">from</span> <span class="nn">._version</span> <span class="k">import</span> <span class="n">TOMBO_VERSION</span>
<span class="kn">from</span> <span class="nn">.c_helper</span> <span class="k">import</span> <span class="n">c_new_mean_stds</span><span class="p">,</span> <span class="n">c_new_means</span><span class="p">,</span> <span class="n">c_apply_outlier_thresh</span>
<span class="kn">from</span> <span class="nn">._default_parameters</span> <span class="k">import</span> <span class="n">ROBUST_QUANTS</span><span class="p">,</span> <span class="n">NUM_READS_FOR_SCALE</span>

<span class="n">VERBOSE</span> <span class="o">=</span> <span class="kc">False</span>


<span class="c1">################################</span>
<span class="c1">###### Global Namedtuples ######</span>
<span class="c1">################################</span>

<span class="n">alignInfo</span> <span class="o">=</span> <span class="n">namedtuple</span><span class="p">(</span>
    <span class="s1">&#39;alignInfo&#39;</span><span class="p">,</span>
    <span class="p">(</span><span class="s1">&#39;ID&#39;</span><span class="p">,</span> <span class="s1">&#39;Subgroup&#39;</span><span class="p">,</span> <span class="s1">&#39;ClipStart&#39;</span><span class="p">,</span> <span class="s1">&#39;ClipEnd&#39;</span><span class="p">,</span>
     <span class="s1">&#39;Insertions&#39;</span><span class="p">,</span> <span class="s1">&#39;Deletions&#39;</span><span class="p">,</span> <span class="s1">&#39;Matches&#39;</span><span class="p">,</span> <span class="s1">&#39;Mismatches&#39;</span><span class="p">))</span>

<span class="n">readData</span> <span class="o">=</span> <span class="n">namedtuple</span><span class="p">(</span><span class="s1">&#39;readData&#39;</span><span class="p">,</span> <span class="p">(</span>
    <span class="s1">&#39;start&#39;</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">,</span> <span class="s1">&#39;filtered&#39;</span><span class="p">,</span> <span class="s1">&#39;read_start_rel_to_raw&#39;</span><span class="p">,</span>
    <span class="s1">&#39;strand&#39;</span><span class="p">,</span> <span class="s1">&#39;fn&#39;</span><span class="p">,</span> <span class="s1">&#39;corr_group&#39;</span><span class="p">,</span> <span class="s1">&#39;rna&#39;</span><span class="p">))</span>

<span class="n">intervalData</span> <span class="o">=</span> <span class="n">namedtuple</span><span class="p">(</span><span class="s1">&#39;intervalData&#39;</span><span class="p">,</span> <span class="p">(</span>
    <span class="s1">&#39;reg_id&#39;</span><span class="p">,</span> <span class="s1">&#39;chrm&#39;</span><span class="p">,</span> <span class="s1">&#39;start&#39;</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">,</span> <span class="s1">&#39;strand&#39;</span><span class="p">,</span> <span class="s1">&#39;reg_text&#39;</span><span class="p">,</span> <span class="s1">&#39;reads&#39;</span><span class="p">,</span> <span class="s1">&#39;seq&#39;</span><span class="p">))</span>
<span class="c1"># set default values for strand, text, reads and seq</span>
<span class="n">intervalData</span><span class="o">.</span><span class="fm">__new__</span><span class="o">.</span><span class="vm">__defaults__</span> <span class="o">=</span> <span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>

<span class="n">channelInfo</span> <span class="o">=</span> <span class="n">namedtuple</span><span class="p">(</span>
    <span class="s1">&#39;channelInfo&#39;</span><span class="p">,</span>
    <span class="p">(</span><span class="s1">&#39;offset&#39;</span><span class="p">,</span> <span class="s1">&#39;range&#39;</span><span class="p">,</span> <span class="s1">&#39;digitisation&#39;</span><span class="p">,</span> <span class="s1">&#39;number&#39;</span><span class="p">,</span> <span class="s1">&#39;sampling_rate&#39;</span><span class="p">))</span>

<span class="n">scaleValues</span> <span class="o">=</span> <span class="n">namedtuple</span><span class="p">(</span>
    <span class="s1">&#39;scaleValues&#39;</span><span class="p">,</span>
    <span class="p">(</span><span class="s1">&#39;shift&#39;</span><span class="p">,</span> <span class="s1">&#39;scale&#39;</span><span class="p">,</span> <span class="s1">&#39;lower_lim&#39;</span><span class="p">,</span> <span class="s1">&#39;upper_lim&#39;</span><span class="p">))</span>

<span class="n">genomeLoc</span> <span class="o">=</span> <span class="n">namedtuple</span><span class="p">(</span>
    <span class="s1">&#39;genomeLoc&#39;</span><span class="p">,</span> <span class="p">(</span><span class="s1">&#39;Start&#39;</span><span class="p">,</span> <span class="s1">&#39;Strand&#39;</span><span class="p">,</span> <span class="s1">&#39;Chrom&#39;</span><span class="p">))</span>

<span class="n">NORM_TYPES</span> <span class="o">=</span> <span class="p">(</span><span class="s1">&#39;none&#39;</span><span class="p">,</span> <span class="s1">&#39;pA&#39;</span><span class="p">,</span> <span class="s1">&#39;pA_raw&#39;</span><span class="p">,</span> <span class="s1">&#39;median&#39;</span><span class="p">,</span> <span class="s1">&#39;robust_median&#39;</span><span class="p">,</span>
              <span class="s1">&#39;median_const_scale&#39;</span><span class="p">)</span>

<span class="c1"># single base conversion for motifs</span>
<span class="n">SINGLE_LETTER_CODE</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;A&#39;</span><span class="p">:</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">:</span><span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">:</span><span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">:</span><span class="s1">&#39;T&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">:</span><span class="s1">&#39;[CGT]&#39;</span><span class="p">,</span>
    <span class="s1">&#39;D&#39;</span><span class="p">:</span><span class="s1">&#39;[AGT]&#39;</span><span class="p">,</span> <span class="s1">&#39;H&#39;</span><span class="p">:</span><span class="s1">&#39;[ACT]&#39;</span><span class="p">,</span> <span class="s1">&#39;K&#39;</span><span class="p">:</span><span class="s1">&#39;[GT]&#39;</span><span class="p">,</span> <span class="s1">&#39;M&#39;</span><span class="p">:</span><span class="s1">&#39;[AC]&#39;</span><span class="p">,</span>
    <span class="s1">&#39;N&#39;</span><span class="p">:</span><span class="s1">&#39;[ACGT]&#39;</span><span class="p">,</span> <span class="s1">&#39;R&#39;</span><span class="p">:</span><span class="s1">&#39;[AG]&#39;</span><span class="p">,</span> <span class="s1">&#39;S&#39;</span><span class="p">:</span><span class="s1">&#39;[CG]&#39;</span><span class="p">,</span> <span class="s1">&#39;V&#39;</span><span class="p">:</span><span class="s1">&#39;[ACG]&#39;</span><span class="p">,</span>
    <span class="s1">&#39;W&#39;</span><span class="p">:</span><span class="s1">&#39;[AT]&#39;</span><span class="p">,</span> <span class="s1">&#39;Y&#39;</span><span class="p">:</span><span class="s1">&#39;[CT]&#39;</span><span class="p">}</span>
<span class="n">INVALID_BASES</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="s1">&#39;[^ACGT]&#39;</span><span class="p">)</span>


<span class="c1">######################################</span>
<span class="c1">###### Various Helper Functions ######</span>
<span class="c1">######################################</span>

<span class="k">def</span> <span class="nf">_warning_message</span><span class="p">(</span><span class="n">message</span><span class="p">):</span>
    <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
        <span class="s1">&#39;*&#39;</span> <span class="o">*</span> <span class="mi">20</span> <span class="o">+</span> <span class="s1">&#39; WARNING &#39;</span> <span class="o">+</span> <span class="s1">&#39;*&#39;</span> <span class="o">*</span> <span class="mi">20</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n\t</span><span class="s1">&#39;</span> <span class="o">+</span>
        <span class="n">message</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="k">return</span>

<span class="k">def</span> <span class="nf">_error_message_and_exit</span><span class="p">(</span><span class="n">message</span><span class="p">):</span>
    <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
        <span class="s1">&#39;*&#39;</span> <span class="o">*</span> <span class="mi">20</span> <span class="o">+</span> <span class="s1">&#39; ERROR &#39;</span> <span class="o">+</span> <span class="s1">&#39;*&#39;</span> <span class="o">*</span> <span class="mi">20</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n\t</span><span class="s1">&#39;</span> <span class="o">+</span>
        <span class="n">message</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">()</span>
    <span class="k">return</span>

<span class="n">COMP_BASES</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">ord</span><span class="p">,</span> <span class="s1">&#39;ACGT&#39;</span><span class="p">),</span> <span class="nb">map</span><span class="p">(</span><span class="nb">ord</span><span class="p">,</span> <span class="s1">&#39;TGCA&#39;</span><span class="p">)))</span>
<div class="viewcode-block" id="comp_seq"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.comp_seq">[docs]</a><span class="k">def</span> <span class="nf">comp_seq</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Complement DNA sequence</span>
<span class="sd">    &quot;&quot;&quot;</span></div>
    <span class="k">return</span> <span class="n">seq</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">COMP_BASES</span><span class="p">)</span>
<div class="viewcode-block" id="rev_comp"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.rev_comp">[docs]</a><span class="k">def</span> <span class="nf">rev_comp</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Reverse complement DNA sequence</span>
<span class="sd">    &quot;&quot;&quot;</span></div>
    <span class="k">return</span> <span class="n">seq</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">COMP_BASES</span><span class="p">)[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

<span class="n">U_TO_T</span> <span class="o">=</span> <span class="p">{</span><span class="nb">ord</span><span class="p">(</span><span class="s1">&#39;U&#39;</span><span class="p">):</span><span class="nb">ord</span><span class="p">(</span><span class="s1">&#39;T&#39;</span><span class="p">)}</span>
<div class="viewcode-block" id="rev_transcribe"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.rev_transcribe">[docs]</a><span class="k">def</span> <span class="nf">rev_transcribe</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Convert U bases to T</span>
<span class="sd">    &quot;&quot;&quot;</span></div>
    <span class="k">return</span> <span class="n">seq</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">U_TO_T</span><span class="p">)</span>

<div class="viewcode-block" id="get_chrm_sizes"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.get_chrm_sizes">[docs]</a><span class="k">def</span> <span class="nf">get_chrm_sizes</span><span class="p">(</span><span class="n">raw_read_coverage</span><span class="p">,</span> <span class="n">raw_read_coverage2</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get covered chromosome sizes from a set of reads</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">strand_chrm_sizes</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">),</span> <span class="n">cs_read_cov</span> <span class="ow">in</span> \
        <span class="n">raw_read_coverage</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">strand_chrm_sizes</span><span class="p">[</span><span class="n">chrm</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span>
            <span class="n">r_data</span><span class="o">.</span><span class="n">end</span> <span class="k">for</span> <span class="n">r_data</span> <span class="ow">in</span> <span class="n">cs_read_cov</span><span class="p">))</span>
    <span class="k">if</span> <span class="n">raw_read_coverage2</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">for</span> <span class="p">(</span><span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">),</span> <span class="n">cs_read_cov</span> <span class="ow">in</span> \
            <span class="n">raw_read_coverage2</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">strand_chrm_sizes</span><span class="p">[</span><span class="n">chrm</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span>
                <span class="n">r_data</span><span class="o">.</span><span class="n">end</span> <span class="k">for</span> <span class="n">r_data</span> <span class="ow">in</span> <span class="n">cs_read_cov</span><span class="p">))</span>

    <span class="k">return</span> <span class="nb">dict</span><span class="p">((</span><span class="n">chrm</span><span class="p">,</span> <span class="nb">max</span><span class="p">(</span><span class="n">strnd_sizes</span><span class="p">))</span>
                <span class="k">for</span> <span class="n">chrm</span><span class="p">,</span> <span class="n">strnd_sizes</span> <span class="ow">in</span></div>
                <span class="n">strand_chrm_sizes</span><span class="o">.</span><span class="n">items</span><span class="p">())</span>

<div class="viewcode-block" id="TomboMotif"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.TomboMotif">[docs]</a><span class="k">class</span> <span class="nc">TomboMotif</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="k">def</span> <span class="nf">_parse_motif</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">rev_comp_motif</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Parse a single letter code motif into a pattern for matching</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">conv_motif</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">SINGLE_LETTER_CODE</span><span class="p">[</span><span class="n">letter</span><span class="p">]</span>
                             <span class="k">for</span> <span class="n">letter</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">raw_motif</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">rev_comp_motif</span><span class="p">:</span>
            <span class="c1"># reverse complement and then flip any group brackets</span>
            <span class="n">conv_motif</span> <span class="o">=</span> <span class="n">rev_comp</span><span class="p">(</span><span class="n">conv_motif</span><span class="p">)</span><span class="o">.</span><span class="n">translate</span><span class="p">({</span>
                <span class="nb">ord</span><span class="p">(</span><span class="s1">&#39;[&#39;</span><span class="p">):</span><span class="s1">&#39;]&#39;</span><span class="p">,</span> <span class="nb">ord</span><span class="p">(</span><span class="s1">&#39;]&#39;</span><span class="p">):</span><span class="s1">&#39;[&#39;</span><span class="p">})</span>
        <span class="k">return</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="n">conv_motif</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">raw_motif</span><span class="p">,</span> <span class="n">mod_pos</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="n">invalid_chars</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span>
            <span class="s1">&#39;[^&#39;</span> <span class="o">+</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">SINGLE_LETTER_CODE</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;]&#39;</span><span class="p">,</span> <span class="n">raw_motif</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">invalid_chars</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">_error_message_and_exit</span><span class="p">(</span>
                <span class="s1">&#39;Invalid characters in motif: &#39;</span> <span class="o">+</span> <span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">invalid_chars</span><span class="p">))</span>

        <span class="c1"># basic motif parsing</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">raw_motif</span> <span class="o">=</span> <span class="n">raw_motif</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">motif_len</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">raw_motif</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">motif_pat</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_parse_motif</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">rev_comp_pat</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_parse_motif</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">is_palindrome</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">motif_pat</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">rev_comp_pat</span>

        <span class="c1"># parse modified position from motif if provided</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mod_pos</span> <span class="o">=</span> <span class="n">mod_pos</span>
        <span class="k">if</span> <span class="n">mod_pos</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">mod_base</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">mod_base</span> <span class="o">=</span> <span class="n">raw_motif</span><span class="p">[</span><span class="n">mod_pos</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">INVALID_BASES</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mod_base</span><span class="p">):</span>
                <span class="n">_warning_message</span><span class="p">(</span>
                    <span class="s1">&#39;Provided modified position is not a single base, which &#39;</span> <span class="o">+</span>
                    <span class="s1">&#39;is likely an error. Specified modified base is one of: &#39;</span> <span class="o">+</span></div>
                    <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">SINGLE_LETTER_CODE</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">mod_base</span><span class="p">][</span><span class="mi">1</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]))</span>


<div class="viewcode-block" id="invalid_seq"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.invalid_seq">[docs]</a><span class="k">def</span> <span class="nf">invalid_seq</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span></div>
    <span class="k">return</span> <span class="nb">bool</span><span class="p">(</span><span class="n">INVALID_BASES</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">seq</span><span class="p">))</span>


<span class="c1">###########################</span>
<span class="c1">###### FASTA Parsing ######</span>
<span class="c1">###########################</span>

<div class="viewcode-block" id="Fasta"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.Fasta">[docs]</a><span class="k">class</span> <span class="nc">Fasta</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Fasta sequence format wrapper class.</span>

<span class="sd">    Will load faidx via pyfaidx package if installed, else the fasta will be</span>
<span class="sd">    loaded into memory for sequence extraction</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">_load_in_mem</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">genome_index</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">curr_id</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">curr_seq</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">with</span> <span class="n">io</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fasta_fn</span><span class="p">)</span> <span class="k">as</span> <span class="n">fasta_fp</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">fasta_fp</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;&gt;&#39;</span><span class="p">):</span>
                    <span class="k">if</span> <span class="p">(</span><span class="n">curr_id</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span>
                        <span class="n">curr_seq</span> <span class="ow">is</span> <span class="ow">not</span> <span class="p">[]):</span>
                        <span class="n">genome_index</span><span class="p">[</span><span class="n">curr_id</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">curr_seq</span><span class="p">)</span>
                    <span class="n">curr_seq</span> <span class="o">=</span> <span class="p">[]</span>
                    <span class="n">curr_id</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;&gt;&quot;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">curr_seq</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">())</span>
            <span class="c1"># add last record</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">curr_id</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span>
                <span class="n">curr_seq</span> <span class="ow">is</span> <span class="ow">not</span> <span class="p">[]):</span>
                <span class="n">genome_index</span><span class="p">[</span><span class="n">curr_id</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">curr_seq</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">genome_index</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fasta_fn</span><span class="p">,</span> <span class="n">dry_run</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">force_in_mem</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fasta_fn</span> <span class="o">=</span> <span class="n">fasta_fn</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">force_in_mem</span><span class="p">:</span> <span class="k">raise</span> <span class="ne">ImportError</span>
            <span class="kn">import</span> <span class="nn">pyfaidx</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">has_pyfaidx</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">pyfaidx</span><span class="o">.</span><span class="n">Fasta</span><span class="p">(</span><span class="n">fasta_fn</span><span class="p">)</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">has_pyfaidx</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">dry_run</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_load_in_mem</span><span class="p">()</span>

<div class="viewcode-block" id="Fasta.get_seq"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.Fasta.get_seq">[docs]</a>    <span class="k">def</span> <span class="nf">get_seq</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chrm</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">has_pyfaidx</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="n">start</span> <span class="ow">or</span> <span class="n">end</span><span class="p">):</span>
                <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">chrm</span><span class="p">][:]</span><span class="o">.</span><span class="n">seq</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">chrm</span><span class="p">][</span><span class="n">start</span><span class="p">:</span><span class="n">end</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span></div>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">chrm</span><span class="p">][</span><span class="n">start</span><span class="p">:</span><span class="n">end</span><span class="p">]</span>

<div class="viewcode-block" id="Fasta.iter_chrms"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.Fasta.iter_chrms">[docs]</a>    <span class="k">def</span> <span class="nf">iter_chrms</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">has_pyfaidx</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">chrm</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">index</span><span class="p">:</span>
                <span class="k">yield</span> <span class="n">unicode</span><span class="p">(</span><span class="n">chrm</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">chrm</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">index</span><span class="p">:</span></div></div>
                <span class="k">yield</span> <span class="n">chrm</span>


<span class="c1">#############################################</span>
<span class="c1">###### Automatic Sample-Type Detection ######</span>
<span class="c1">#############################################</span>

<div class="viewcode-block" id="is_read_rna"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.is_read_rna">[docs]</a><span class="k">def</span> <span class="nf">is_read_rna</span><span class="p">(</span><span class="n">fast5_data</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Determine if a read is RNA or DNA</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># check both experiment type and kit slots for &quot;rna&quot;</span>
    <span class="n">exp_type</span><span class="p">,</span> <span class="n">exp_kit</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">exp_type</span> <span class="o">=</span> <span class="n">fast5_data</span><span class="p">[</span><span class="s1">&#39;UniqueGlobalKey/context_tags&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span>
            <span class="s1">&#39;experiment_type&#39;</span><span class="p">]</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">exp_type</span> <span class="o">=</span> <span class="n">exp_type</span><span class="o">.</span><span class="n">decode</span><span class="p">()</span>
        <span class="k">except</span> <span class="p">(</span><span class="ne">AttributeError</span><span class="p">,</span> <span class="ne">TypeError</span><span class="p">):</span>
            <span class="k">pass</span>
        <span class="c1"># remove the word internal since it contains rna.</span>
        <span class="n">exp_type</span> <span class="o">=</span> <span class="n">exp_type</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;internal&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="k">pass</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">exp_kit</span> <span class="o">=</span> <span class="n">fast5_data</span><span class="p">[</span><span class="s1">&#39;UniqueGlobalKey/context_tags&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span>
            <span class="s1">&#39;experiment_kit&#39;</span><span class="p">]</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">exp_kit</span> <span class="o">=</span> <span class="n">exp_kit</span><span class="o">.</span><span class="n">decode</span><span class="p">()</span>
        <span class="k">except</span> <span class="p">(</span><span class="ne">AttributeError</span><span class="p">,</span> <span class="ne">TypeError</span><span class="p">):</span>
            <span class="k">pass</span>
        <span class="c1"># remove the word internal since it contains rna.</span>
        <span class="n">exp_kit</span> <span class="o">=</span> <span class="n">exp_kit</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;internal&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="k">pass</span>

    <span class="k">if</span> <span class="n">exp_type</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">exp_kit</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">rna</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">rna</span> <span class="o">=</span> <span class="p">(</span>
            <span class="p">(</span><span class="n">exp_type</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="s1">&#39;rna&#39;</span><span class="p">,</span> <span class="n">exp_type</span><span class="p">)</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">or</span>
            <span class="p">(</span><span class="n">exp_kit</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="s1">&#39;rna&#39;</span><span class="p">,</span> <span class="n">exp_kit</span><span class="p">)</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">))</span>
</div>
    <span class="k">return</span> <span class="n">rna</span>

<div class="viewcode-block" id="is_rna"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.is_rna">[docs]</a><span class="k">def</span> <span class="nf">is_rna</span><span class="p">(</span><span class="n">raw_read_coverage</span><span class="p">,</span> <span class="n">n_reads</span><span class="o">=</span><span class="mi">10</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Determine if a set of reads are RNA or DNA from a small sample</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">proc_reads</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">cs_reads</span> <span class="ow">in</span> <span class="n">raw_read_coverage</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
        <span class="k">for</span> <span class="n">r_data</span> <span class="ow">in</span> <span class="n">cs_reads</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">r_data</span><span class="o">.</span><span class="n">rna</span><span class="p">:</span>
                <span class="k">return</span> <span class="kc">False</span>
            <span class="n">proc_reads</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">if</span> <span class="n">proc_reads</span> <span class="o">&gt;=</span> <span class="n">n_reads</span><span class="p">:</span>
                <span class="k">break</span>
        <span class="k">if</span> <span class="n">proc_reads</span> <span class="o">&gt;=</span> <span class="n">n_reads</span><span class="p">:</span>
            <span class="k">break</span></div>
    <span class="k">return</span> <span class="kc">True</span>

<div class="viewcode-block" id="is_rna_from_files"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.is_rna_from_files">[docs]</a><span class="k">def</span> <span class="nf">is_rna_from_files</span><span class="p">(</span><span class="n">fast5_fns</span><span class="p">,</span> <span class="n">n_reads</span><span class="o">=</span><span class="mi">10</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Determine if a set of files are RNA or DNA from a small sample</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">proc_reads</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">fast5_fn</span> <span class="ow">in</span> <span class="n">fast5_fns</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">fast5_data</span> <span class="o">=</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="n">fast5_fn</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">is_read_rna</span><span class="p">(</span><span class="n">fast5_data</span><span class="p">):</span>
                <span class="k">return</span> <span class="kc">False</span>
            <span class="n">proc_reads</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="k">if</span> <span class="n">proc_reads</span> <span class="o">&gt;=</span> <span class="n">n_reads</span><span class="p">:</span>
            <span class="k">break</span></div>
    <span class="k">return</span> <span class="kc">True</span>


<span class="c1">#########################################</span>
<span class="c1">###### Index File/Filter Functions ######</span>
<span class="c1">#########################################</span>

<div class="viewcode-block" id="get_index_fn"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.get_index_fn">[docs]</a><span class="k">def</span> <span class="nf">get_index_fn</span><span class="p">(</span><span class="n">fast5s_dir</span><span class="p">,</span> <span class="n">corr_grp</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get the filename for the requested directory and corrected group</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># dirname should come in with a trailing slash</span>
    <span class="n">split_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">fast5s_dir</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
    <span class="k">return</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">split_dir</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="s2">&quot;.&quot;</span> <span class="o">+</span> <span class="n">split_dir</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span></div>
                        <span class="s2">&quot;.&quot;</span> <span class="o">+</span> <span class="n">corr_grp</span> <span class="o">+</span> <span class="s1">&#39;.tombo.index&#39;</span><span class="p">)</span>

<div class="viewcode-block" id="prep_index_data"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.prep_index_data">[docs]</a><span class="k">def</span> <span class="nf">prep_index_data</span><span class="p">(</span>
        <span class="n">fast5_fn</span><span class="p">,</span> <span class="n">genome_loc</span><span class="p">,</span> <span class="n">read_start_rel_to_raw</span><span class="p">,</span> <span class="n">segs</span><span class="p">,</span>
        <span class="n">corrected_group</span><span class="p">,</span> <span class="n">subgroup</span><span class="p">,</span> <span class="n">rna</span><span class="p">,</span> <span class="n">obs_filter</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Prepare data for storage in the index file</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">obs_filter</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">is_filtered</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">base_lens</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">diff</span><span class="p">(</span><span class="n">segs</span><span class="p">)</span>
        <span class="n">is_filtered</span> <span class="o">=</span> <span class="nb">any</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">base_lens</span><span class="p">,</span> <span class="n">pctl</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">thresh</span>
                          <span class="k">for</span> <span class="n">pctl</span><span class="p">,</span> <span class="n">thresh</span> <span class="ow">in</span> <span class="n">obs_filter</span><span class="p">)</span>
    <span class="n">mapped_end</span> <span class="o">=</span> <span class="n">genome_loc</span><span class="o">.</span><span class="n">Start</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">segs</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span>

    <span class="k">return</span> <span class="p">((</span><span class="n">genome_loc</span><span class="o">.</span><span class="n">Chrom</span><span class="p">,</span> <span class="n">genome_loc</span><span class="o">.</span><span class="n">Strand</span><span class="p">),</span> <span class="p">(</span>
        <span class="n">fast5_fn</span><span class="p">,</span> <span class="n">genome_loc</span><span class="o">.</span><span class="n">Start</span><span class="p">,</span> <span class="n">mapped_end</span><span class="p">,</span> <span class="n">read_start_rel_to_raw</span><span class="p">,</span></div>
        <span class="n">corrected_group</span><span class="p">,</span> <span class="n">subgroup</span><span class="p">,</span> <span class="n">is_filtered</span><span class="p">,</span> <span class="n">rna</span><span class="p">))</span>

<div class="viewcode-block" id="write_index_file"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.write_index_file">[docs]</a><span class="k">def</span> <span class="nf">write_index_file</span><span class="p">(</span><span class="n">all_index_data</span><span class="p">,</span> <span class="n">index_fn</span><span class="p">,</span> <span class="n">basedir</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Write index file</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="kn">import</span> <span class="nn">cPickle</span> <span class="k">as</span> <span class="nn">pickle</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="kn">import</span> <span class="nn">pickle</span>
    <span class="n">index_data</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">chrm_strand</span><span class="p">,</span> <span class="p">(</span><span class="n">fn</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">rsrtr</span><span class="p">,</span> <span class="n">c_grp</span><span class="p">,</span> <span class="n">s_grp</span><span class="p">,</span>
                      <span class="n">filtered</span><span class="p">,</span> <span class="n">rna</span><span class="p">)</span> <span class="ow">in</span> <span class="n">all_index_data</span><span class="p">:</span>
        <span class="c1"># clip the basedir off the FAST5 filename in case later functions are</span>
        <span class="c1"># called from another relative path</span>
        <span class="n">from_base_fn</span> <span class="o">=</span> <span class="n">fn</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">basedir</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span>
        <span class="n">index_data</span><span class="p">[</span><span class="n">chrm_strand</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">((</span>
            <span class="n">from_base_fn</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">rsrtr</span><span class="p">,</span> <span class="n">c_grp</span><span class="p">,</span> <span class="n">s_grp</span><span class="p">,</span> <span class="n">filtered</span><span class="p">,</span> <span class="n">rna</span><span class="p">))</span>

    <span class="k">with</span> <span class="n">io</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">index_fn</span><span class="p">,</span> <span class="s1">&#39;wb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">index_fp</span><span class="p">:</span>
        <span class="c1"># note protocol 2 for py2/3 compatibility</span>
        <span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="nb">dict</span><span class="p">(</span><span class="n">index_data</span><span class="p">),</span> <span class="n">index_fp</span><span class="p">,</span> <span class="n">protocol</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
</div>
    <span class="k">return</span>

<div class="viewcode-block" id="clear_filters"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.clear_filters">[docs]</a><span class="k">def</span> <span class="nf">clear_filters</span><span class="p">(</span><span class="n">fast5s_dir</span><span class="p">,</span> <span class="n">corr_grp</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Clear filters applied to this directories index files</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">fast5s_dir</span> <span class="o">=</span> <span class="p">(</span><span class="n">fast5s_dir</span> <span class="k">if</span> <span class="n">fast5s_dir</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)</span> <span class="k">else</span>
                  <span class="n">fast5s_dir</span> <span class="o">+</span> <span class="s1">&#39;/&#39;</span><span class="p">)</span>
    <span class="n">index_fn</span> <span class="o">=</span> <span class="n">get_index_fn</span><span class="p">(</span><span class="n">fast5s_dir</span><span class="p">,</span> <span class="n">corr_grp</span><span class="p">)</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="kn">import</span> <span class="nn">cPickle</span> <span class="k">as</span> <span class="nn">pickle</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="kn">import</span> <span class="nn">pickle</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="k">with</span> <span class="n">io</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">index_fn</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">index_fp</span><span class="p">:</span>
            <span class="n">index_data</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">index_fp</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
        <span class="n">_error_message_and_exit</span><span class="p">(</span>
            <span class="s1">&#39;Filters can only be applied to runs &#39;</span> <span class="o">+</span>
            <span class="s1">&#39;with a Tombo index file. Re-run resquiggle without the &#39;</span> <span class="o">+</span>
            <span class="s1">&#39;--skip-index option to apply filters.&#39;</span><span class="p">)</span>
    <span class="n">new_index_data</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">chrm_strand</span><span class="p">,</span> <span class="n">cs_raw_data</span> <span class="ow">in</span> <span class="n">index_data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">new_index_data</span><span class="o">.</span><span class="n">extend</span><span class="p">([(</span><span class="n">chrm_strand</span><span class="p">,</span> <span class="p">(</span>
            <span class="n">from_base_fn</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">rsrtr</span><span class="p">,</span> <span class="n">corr_grp</span><span class="p">,</span> <span class="n">s_grp</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="n">rna</span><span class="p">))</span>
                               <span class="k">for</span> <span class="n">from_base_fn</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">rsrtr</span><span class="p">,</span> <span class="n">c_grp</span><span class="p">,</span>
                               <span class="n">s_grp</span><span class="p">,</span> <span class="n">filtered</span><span class="p">,</span> <span class="n">rna</span> <span class="ow">in</span> <span class="n">cs_raw_data</span><span class="p">])</span>

    <span class="n">write_index_file</span><span class="p">(</span><span class="n">new_index_data</span><span class="p">,</span> <span class="n">index_fn</span><span class="p">,</span> <span class="n">fast5s_dir</span><span class="p">)</span>
    <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;All filters successfully cleared!</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
</div>
    <span class="k">return</span>

<div class="viewcode-block" id="parse_obs_filter"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.parse_obs_filter">[docs]</a><span class="k">def</span> <span class="nf">parse_obs_filter</span><span class="p">(</span><span class="n">obs_filter</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Parse observations per base formatted filtering</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">obs_filter</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="c1"># parse obs_filter</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">obs_filter</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">int</span><span class="p">,</span> <span class="n">pctl_nobs</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;:&#39;</span><span class="p">)))</span>
                      <span class="k">for</span> <span class="n">pctl_nobs</span> <span class="ow">in</span> <span class="n">obs_filter</span><span class="p">]</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="s1">&#39;Invalid format for observation filter&#39;</span><span class="p">)</span>

    <span class="k">if</span> <span class="nb">any</span><span class="p">(</span><span class="n">pctl</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">pctl</span> <span class="o">&gt;</span> <span class="mi">100</span> <span class="k">for</span> <span class="n">pctl</span> <span class="ow">in</span> <span class="nb">map</span><span class="p">(</span><span class="n">itemgetter</span><span class="p">(</span><span class="mi">0</span><span class="p">),</span> <span class="n">obs_filter</span><span class="p">)):</span>
       <span class="n">_error_message_and_exit</span><span class="p">(</span><span class="s1">&#39;Invalid percentile value.&#39;</span><span class="p">)</span>
</div>
    <span class="k">return</span> <span class="n">obs_filter</span>

<div class="viewcode-block" id="filter_reads"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.filter_reads">[docs]</a><span class="k">def</span> <span class="nf">filter_reads</span><span class="p">(</span><span class="n">fast5s_dir</span><span class="p">,</span> <span class="n">corr_grp</span><span class="p">,</span> <span class="n">obs_filter</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Filter reads based on some observation per base threshold criteria</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">read_is_stuck</span><span class="p">(</span><span class="n">fast5_fn</span><span class="p">,</span> <span class="n">s_grp</span><span class="p">):</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">fast5_data</span> <span class="o">=</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="n">fast5_fn</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span>
            <span class="n">base_lens</span> <span class="o">=</span> <span class="n">fast5_data</span><span class="p">[</span><span class="s1">&#39;/Analyses/&#39;</span> <span class="o">+</span> <span class="n">corr_grp</span> <span class="o">+</span> <span class="s1">&#39;/&#39;</span> <span class="o">+</span> <span class="n">s_grp</span> <span class="o">+</span>
                                   <span class="s1">&#39;/Events&#39;</span><span class="p">][</span><span class="s1">&#39;length&#39;</span><span class="p">]</span>
            <span class="k">return</span> <span class="nb">any</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">base_lens</span><span class="p">,</span> <span class="n">pctl</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">thresh</span>
                       <span class="k">for</span> <span class="n">pctl</span><span class="p">,</span> <span class="n">thresh</span> <span class="ow">in</span> <span class="n">obs_filter</span><span class="p">)</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">True</span>

    <span class="n">fast5s_dir</span> <span class="o">=</span> <span class="p">(</span><span class="n">fast5s_dir</span> <span class="k">if</span> <span class="n">fast5s_dir</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)</span> <span class="k">else</span>
                  <span class="n">fast5s_dir</span> <span class="o">+</span> <span class="s1">&#39;/&#39;</span><span class="p">)</span>
    <span class="n">index_fn</span> <span class="o">=</span> <span class="n">get_index_fn</span><span class="p">(</span><span class="n">fast5s_dir</span><span class="p">,</span> <span class="n">corr_grp</span><span class="p">)</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="kn">import</span> <span class="nn">cPickle</span> <span class="k">as</span> <span class="nn">pickle</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="kn">import</span> <span class="nn">pickle</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="k">with</span> <span class="n">io</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">index_fn</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">index_fp</span><span class="p">:</span>
            <span class="n">index_data</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">index_fp</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
            <span class="s1">&#39;******** ERRROR *******</span><span class="se">\n\t</span><span class="s1">Filters can only be applied to runs &#39;</span> <span class="o">+</span>
            <span class="s1">&#39;with a Tombo index file. Re-run resquiggle without the &#39;</span> <span class="o">+</span>
            <span class="s1">&#39;--skip-index option to apply filters.</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="n">filt_index_data</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">num_reads</span><span class="p">,</span> <span class="n">num_filt_reads</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">chrm_strand</span><span class="p">,</span> <span class="n">cs_raw_data</span> <span class="ow">in</span> <span class="n">index_data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">cs_filt_reads</span> <span class="o">=</span> <span class="p">[</span>
            <span class="p">(</span><span class="n">chrm_strand</span><span class="p">,</span> <span class="p">(</span>
                <span class="n">from_base_fn</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">rsrtr</span><span class="p">,</span> <span class="n">corr_grp</span><span class="p">,</span> <span class="n">s_grp</span><span class="p">,</span>
                <span class="n">read_is_stuck</span><span class="p">(</span><span class="n">fast5s_dir</span> <span class="o">+</span> <span class="s1">&#39;/&#39;</span> <span class="o">+</span> <span class="n">from_base_fn</span><span class="p">,</span> <span class="n">s_grp</span><span class="p">),</span> <span class="n">rna</span><span class="p">))</span>
            <span class="k">for</span> <span class="n">from_base_fn</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">rsrtr</span><span class="p">,</span> <span class="n">c_grp</span><span class="p">,</span>
            <span class="n">s_grp</span><span class="p">,</span> <span class="n">filtered</span><span class="p">,</span> <span class="n">rna</span> <span class="ow">in</span> <span class="n">cs_raw_data</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">filtered</span><span class="p">]</span>
        <span class="n">num_reads</span> <span class="o">+=</span> <span class="nb">len</span><span class="p">(</span><span class="n">cs_raw_data</span><span class="p">)</span>
        <span class="n">num_filt_reads</span> <span class="o">+=</span> <span class="nb">sum</span><span class="p">([</span><span class="mi">1</span> <span class="k">for</span> <span class="n">i_data</span> <span class="ow">in</span> <span class="n">cs_filt_reads</span> <span class="k">if</span> <span class="n">i_data</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="mi">6</span><span class="p">]])</span>
        <span class="n">filt_index_data</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">cs_filt_reads</span><span class="p">)</span>

    <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
        <span class="s1">&#39;Filtered &#39;</span> <span class="o">+</span> <span class="n">unicode</span><span class="p">(</span><span class="n">num_filt_reads</span><span class="p">)</span> <span class="o">+</span>
        <span class="s1">&#39; reads due to observations per base filter from a &#39;</span> <span class="o">+</span>
        <span class="s1">&#39;total of &#39;</span> <span class="o">+</span> <span class="n">unicode</span><span class="p">(</span><span class="n">num_reads</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39; reads in &#39;</span> <span class="o">+</span> <span class="n">fast5s_dir</span> <span class="o">+</span> <span class="s1">&#39;.</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="n">write_index_file</span><span class="p">(</span><span class="n">filt_index_data</span><span class="p">,</span> <span class="n">index_fn</span><span class="p">,</span> <span class="n">fast5s_dir</span><span class="p">)</span>
</div>
    <span class="k">return</span>

<div class="viewcode-block" id="filter_reads_for_coverage"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.filter_reads_for_coverage">[docs]</a><span class="k">def</span> <span class="nf">filter_reads_for_coverage</span><span class="p">(</span><span class="n">fast5s_dir</span><span class="p">,</span> <span class="n">corr_grp</span><span class="p">,</span> <span class="n">frac_to_filter</span><span class="p">):</span>
    <span class="n">fast5s_dir</span> <span class="o">=</span> <span class="p">(</span><span class="n">fast5s_dir</span> <span class="k">if</span> <span class="n">fast5s_dir</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)</span> <span class="k">else</span>
                  <span class="n">fast5s_dir</span> <span class="o">+</span> <span class="s1">&#39;/&#39;</span><span class="p">)</span>
    <span class="n">index_fn</span> <span class="o">=</span> <span class="n">get_index_fn</span><span class="p">(</span><span class="n">fast5s_dir</span><span class="p">,</span> <span class="n">corr_grp</span><span class="p">)</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="kn">import</span> <span class="nn">cPickle</span> <span class="k">as</span> <span class="nn">pickle</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="kn">import</span> <span class="nn">pickle</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="k">with</span> <span class="n">io</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">index_fn</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">index_fp</span><span class="p">:</span>
            <span class="n">index_data</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">index_fp</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
            <span class="s1">&#39;******** ERRROR *******</span><span class="se">\n\t</span><span class="s1">Filters can only be applied to runs &#39;</span> <span class="o">+</span>
            <span class="s1">&#39;with a Tombo index file. Re-run resquiggle without the &#39;</span> <span class="o">+</span>
            <span class="s1">&#39;--skip-index option to apply filters.</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="n">unfilt_data</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">unfilt_reads_cov</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">prev_filt_data</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">chrm_strand</span><span class="p">,</span> <span class="n">cs_raw_data</span> <span class="ow">in</span> <span class="n">index_data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">max_end</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">end</span> <span class="k">for</span> <span class="p">(</span><span class="n">_</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span><span class="p">)</span> <span class="ow">in</span> <span class="n">cs_raw_data</span><span class="p">)</span>
        <span class="n">cs_coverage</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">max_end</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int64</span><span class="p">)</span>
        <span class="k">for</span> <span class="p">(</span><span class="n">from_base_fn</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">rsrtr</span><span class="p">,</span> <span class="n">c_grp</span><span class="p">,</span>
             <span class="n">s_grp</span><span class="p">,</span> <span class="n">filtered</span><span class="p">,</span> <span class="n">rna</span><span class="p">)</span> <span class="ow">in</span> <span class="n">cs_raw_data</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">filtered</span><span class="p">:</span> <span class="k">continue</span>
            <span class="n">cs_coverage</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">end</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="c1"># now go through and compute coverage as well</span>
        <span class="k">for</span> <span class="p">(</span><span class="n">from_base_fn</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">rsrtr</span><span class="p">,</span> <span class="n">c_grp</span><span class="p">,</span>
             <span class="n">s_grp</span><span class="p">,</span> <span class="n">filtered</span><span class="p">,</span> <span class="n">rna</span><span class="p">)</span> <span class="ow">in</span> <span class="n">cs_raw_data</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">filtered</span><span class="p">:</span>
                <span class="n">prev_filt_data</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">chrm_strand</span><span class="p">,</span> <span class="p">(</span>
                    <span class="n">from_base_fn</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">rsrtr</span><span class="p">,</span>
                    <span class="n">c_grp</span><span class="p">,</span> <span class="n">s_grp</span><span class="p">,</span> <span class="n">filtered</span><span class="p">,</span> <span class="n">rna</span><span class="p">)))</span>
                <span class="k">continue</span>
            <span class="c1"># add approximate coverage from middle of read</span>
            <span class="c1"># faster than mean over the whole read</span>
            <span class="n">unfilt_reads_cov</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cs_coverage</span><span class="p">[</span><span class="n">start</span> <span class="o">+</span> <span class="p">((</span><span class="n">end</span> <span class="o">-</span> <span class="n">start</span><span class="p">)</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)])</span>
            <span class="n">unfilt_data</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">chrm_strand</span><span class="p">,</span> <span class="p">(</span>
                <span class="n">from_base_fn</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">rsrtr</span><span class="p">,</span> <span class="n">c_grp</span><span class="p">,</span> <span class="n">s_grp</span><span class="p">,</span> <span class="n">filtered</span><span class="p">,</span> <span class="n">rna</span><span class="p">)))</span>

    <span class="n">num_reads</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">unfilt_data</span><span class="p">)</span>
    <span class="n">num_filt_reads</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">frac_to_filter</span> <span class="o">*</span> <span class="n">num_reads</span><span class="p">)</span>
    <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
        <span class="s1">&#39;Filtered &#39;</span> <span class="o">+</span> <span class="n">unicode</span><span class="p">(</span><span class="n">num_filt_reads</span><span class="p">)</span> <span class="o">+</span>
        <span class="s1">&#39; reads due to observations per base filter from a &#39;</span> <span class="o">+</span>
        <span class="s1">&#39;total of &#39;</span> <span class="o">+</span> <span class="n">unicode</span><span class="p">(</span><span class="n">num_reads</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39; reads in &#39;</span> <span class="o">+</span> <span class="n">fast5s_dir</span> <span class="o">+</span> <span class="s1">&#39;.</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="c1"># create probabilities array with coverage values normalized to sum to 1</span>
    <span class="n">unfilt_reads_cov</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">unfilt_reads_cov</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float</span><span class="p">)</span>
    <span class="n">unfilt_reads_p</span> <span class="o">=</span> <span class="n">unfilt_reads_cov</span> <span class="o">/</span> <span class="n">unfilt_reads_cov</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
    <span class="n">filt_indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span>
        <span class="n">num_reads</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">num_filt_reads</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">p</span><span class="o">=</span><span class="n">unfilt_reads_p</span><span class="p">)</span>
    <span class="n">filt_index_data</span> <span class="o">=</span> <span class="p">[</span>
        <span class="p">(</span><span class="n">chrm_strand</span><span class="p">,</span> <span class="p">(</span><span class="n">from_base_fn</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">rsrtr</span><span class="p">,</span> <span class="n">c_grp</span><span class="p">,</span> <span class="n">s_grp</span><span class="p">,</span> <span class="kc">True</span><span class="p">,</span> <span class="n">rna</span><span class="p">))</span>
        <span class="k">for</span> <span class="p">(</span><span class="n">chrm_strand</span><span class="p">,</span> <span class="p">(</span><span class="n">from_base_fn</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">rsrtr</span><span class="p">,</span> <span class="n">c_grp</span><span class="p">,</span> <span class="n">s_grp</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">rna</span><span class="p">))</span>
        <span class="ow">in</span> <span class="n">itemgetter</span><span class="p">(</span><span class="o">*</span><span class="n">filt_indices</span><span class="p">)(</span><span class="n">unfilt_data</span><span class="p">)]</span>
    <span class="n">unfilt_index_data</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">itemgetter</span><span class="p">(</span><span class="o">*</span><span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">num_reads</span><span class="p">))</span><span class="o">.</span><span class="n">difference</span><span class="p">(</span>
        <span class="n">filt_indices</span><span class="p">)))(</span><span class="n">unfilt_data</span><span class="p">))</span>

    <span class="n">write_index_file</span><span class="p">(</span><span class="n">prev_filt_data</span> <span class="o">+</span> <span class="n">filt_index_data</span> <span class="o">+</span> <span class="n">unfilt_index_data</span><span class="p">,</span>
                     <span class="n">index_fn</span><span class="p">,</span> <span class="n">fast5s_dir</span><span class="p">)</span>
</div>
    <span class="k">return</span>


<span class="c1">#####################################</span>
<span class="c1">###### FAST5 Parsing Functions ######</span>
<span class="c1">#####################################</span>

<div class="viewcode-block" id="annotate_with_fastqs"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.annotate_with_fastqs">[docs]</a><span class="k">def</span> <span class="nf">annotate_with_fastqs</span><span class="p">(</span><span class="n">fastq_fns</span><span class="p">,</span> <span class="n">fast5s_read_ids</span><span class="p">,</span> <span class="n">fastq_slot</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">VERBOSE</span><span class="p">:</span> <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;Annotating FAST5s with sequence from FASTQs.</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">fastq_fn</span> <span class="ow">in</span> <span class="n">fastq_fns</span><span class="p">:</span>
        <span class="n">n_recs</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">been_warned_ids</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">with</span> <span class="n">io</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">fastq_fn</span><span class="p">)</span> <span class="k">as</span> <span class="n">fastq_fp</span><span class="p">:</span>
            <span class="k">while</span> <span class="kc">True</span><span class="p">:</span>
                <span class="n">fastq_rec</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">islice</span><span class="p">(</span><span class="n">fastq_fp</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>
                <span class="c1"># if record contains fewer than 4 lines this indicates the</span>
                <span class="c1"># EOF, so move to next file</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">fastq_rec</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">4</span><span class="p">:</span> <span class="k">break</span>
                <span class="c1"># if sequence identifier line does not start with &quot;@&quot; or quality</span>
                <span class="c1"># score line does not start with a &quot;+&quot; the file may be</span>
                <span class="c1"># corrupted, so don&#39;t process any more records</span>
                <span class="k">if</span> <span class="p">(</span><span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="s1">&#39;@&#39;</span><span class="p">,</span> <span class="n">fastq_rec</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span>
                    <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="s1">&#39;\+&#39;</span><span class="p">,</span> <span class="n">fastq_rec</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">):</span>
                    <span class="n">_warning_message</span><span class="p">(</span>
                        <span class="s1">&#39;Successfully parsed &#39;</span> <span class="o">+</span> <span class="n">unicode</span><span class="p">(</span><span class="n">n_recs</span><span class="p">)</span> <span class="o">+</span>
                        <span class="s1">&#39;FASTQ records from &#39;</span> <span class="o">+</span> <span class="n">fastq_fn</span> <span class="o">+</span> <span class="s1">&#39; before &#39;</span> <span class="o">+</span>
                        <span class="s1">&#39;encountering an invalid record. The rest of &#39;</span> <span class="o">+</span>
                        <span class="s1">&#39;this file will not be processed.&#39;</span><span class="p">)</span>
                    <span class="k">break</span>

                <span class="c1"># extract read_id from fastq (which should be the first text</span>
                <span class="c1"># after the &quot;@&quot; record delimiter up to the first white space or</span>
                <span class="c1"># underscore</span>
                <span class="n">read_id</span> <span class="o">=</span> <span class="n">fastq_rec</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">:]</span>
                <span class="k">if</span> <span class="n">read_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">fast5s_read_ids</span><span class="p">:</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">been_warned_ids</span><span class="p">:</span>
                        <span class="n">been_warned_ids</span> <span class="o">=</span> <span class="kc">True</span>
                        <span class="n">_warning_message</span><span class="p">(</span>
                            <span class="s1">&#39;Some records from &#39;</span> <span class="o">+</span> <span class="n">fastq_fn</span> <span class="o">+</span> <span class="s1">&#39; contain read &#39;</span> <span class="o">+</span>
                            <span class="s1">&#39;identifiers not found in any FAST5 files.&#39;</span><span class="p">)</span>
                    <span class="k">continue</span>

                <span class="k">with</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="n">fast5s_read_ids</span><span class="p">[</span><span class="n">read_id</span><span class="p">])</span> <span class="k">as</span> <span class="n">fast5_data</span><span class="p">:</span>
                    <span class="n">bc_slot</span> <span class="o">=</span> <span class="n">fast5_data</span><span class="p">[</span><span class="n">fastq_slot</span><span class="p">]</span>
                    <span class="n">bc_slot</span><span class="o">.</span><span class="n">create_dataset</span><span class="p">(</span>
                        <span class="s1">&#39;Fastq&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">fastq_rec</span><span class="p">),</span>
                        <span class="n">dtype</span><span class="o">=</span><span class="n">h5py</span><span class="o">.</span><span class="n">special_dtype</span><span class="p">(</span><span class="n">vlen</span><span class="o">=</span><span class="n">unicode</span><span class="p">))</span>
</div>
    <span class="k">return</span>

<div class="viewcode-block" id="reads_contain_basecalls"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.reads_contain_basecalls">[docs]</a><span class="k">def</span> <span class="nf">reads_contain_basecalls</span><span class="p">(</span><span class="n">fast5_fns</span><span class="p">,</span> <span class="n">bc_grp</span><span class="p">,</span> <span class="n">num_reads</span><span class="p">):</span>
    <span class="n">test_fns</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span>
        <span class="n">fast5_fns</span><span class="p">,</span> <span class="n">num_reads</span><span class="p">)</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">fast5_fns</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">num_reads</span> <span class="k">else</span> <span class="n">fast5_fns</span>
    <span class="k">for</span> <span class="n">fast5_fn</span> <span class="ow">in</span> <span class="n">test_fns</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">with</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="n">fast5_fn</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fast5_data</span><span class="p">:</span>
                <span class="n">fast5_data</span><span class="p">[</span><span class="s1">&#39;/Analyses/&#39;</span> <span class="o">+</span> <span class="n">bc_grp</span><span class="p">]</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="c1"># if the basecall group is accessible for a single file return true</span>
        <span class="k">return</span> <span class="kc">True</span>

    <span class="c1"># else if all tested reads did not contain the basecall group return False</span></div>
    <span class="k">return</span> <span class="kc">False</span>

<div class="viewcode-block" id="get_files_list"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.get_files_list">[docs]</a><span class="k">def</span> <span class="nf">get_files_list</span><span class="p">(</span><span class="n">fast5s_dir</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get all fast5 files recursively listed below the directory</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">all_fast5s</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="c1"># walk through directory structure searching for fast5 files</span>
    <span class="k">for</span> <span class="n">root</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">fns</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">walk</span><span class="p">(</span><span class="n">fast5s_dir</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">fn</span> <span class="ow">in</span> <span class="n">fnmatch</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">fns</span><span class="p">,</span> <span class="s1">&#39;*.fast5&#39;</span><span class="p">):</span>
            <span class="n">all_fast5s</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">root</span><span class="p">,</span> <span class="n">fn</span><span class="p">))</span>
</div>
    <span class="k">return</span> <span class="n">all_fast5s</span>

<div class="viewcode-block" id="get_raw_read_slot"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.get_raw_read_slot">[docs]</a><span class="k">def</span> <span class="nf">get_raw_read_slot</span><span class="p">(</span><span class="n">fast5_data</span><span class="p">):</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">raw_read_slot</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">fast5_data</span><span class="p">[</span><span class="s1">&#39;/Raw/Reads&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">())[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span>
            <span class="s1">&#39;Raw data is not found in /Raw/Reads/Read_[read#]&#39;</span><span class="p">)</span>
</div>
    <span class="k">return</span> <span class="n">raw_read_slot</span>

<div class="viewcode-block" id="prep_fast5_for_fastq"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.prep_fast5_for_fastq">[docs]</a><span class="k">def</span> <span class="nf">prep_fast5_for_fastq</span><span class="p">(</span>
        <span class="n">fast5_data</span><span class="p">,</span> <span class="n">basecall_group</span><span class="p">,</span> <span class="n">basecall_subgroup</span><span class="p">,</span> <span class="n">overwrite</span><span class="p">):</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">read_id</span> <span class="o">=</span> <span class="n">get_raw_read_slot</span><span class="p">(</span><span class="n">fast5_data</span><span class="p">)</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s1">&#39;read_id&#39;</span><span class="p">]</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">read_id</span> <span class="o">=</span> <span class="n">read_id</span><span class="o">.</span><span class="n">decode</span><span class="p">()</span>
        <span class="k">except</span> <span class="p">(</span><span class="ne">AttributeError</span><span class="p">,</span> <span class="ne">TypeError</span><span class="p">):</span>
            <span class="k">pass</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="c1"># if Analyses group doesn&#39;t exist yet, create it</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">analyses_grp</span> <span class="o">=</span> <span class="n">fast5_data</span><span class="p">[</span><span class="s1">&#39;/Analyses&#39;</span><span class="p">]</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="n">analyses_grp</span> <span class="o">=</span> <span class="n">fast5_data</span><span class="o">.</span><span class="n">create_group</span><span class="p">(</span><span class="s1">&#39;Analyses&#39;</span><span class="p">)</span>

    <span class="c1"># create Fastq slot, unless value exists and --overwrite is not set</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">bc_grp</span> <span class="o">=</span> <span class="n">analyses_grp</span><span class="p">[</span><span class="n">basecall_group</span><span class="p">]</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="n">bc_grp</span> <span class="o">=</span> <span class="n">analyses_grp</span><span class="o">.</span><span class="n">create_group</span><span class="p">(</span><span class="n">basecall_group</span><span class="p">)</span>
        <span class="n">bc_subgrp</span> <span class="o">=</span> <span class="n">bc_grp</span><span class="o">.</span><span class="n">create_group</span><span class="p">(</span><span class="n">basecall_subgroup</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">overwrite</span><span class="p">:</span>
            <span class="k">del</span> <span class="n">analyses_grp</span><span class="p">[</span><span class="n">basecall_group</span><span class="p">]</span>
            <span class="n">bc_grp</span> <span class="o">=</span> <span class="n">analyses_grp</span><span class="o">.</span><span class="n">create_group</span><span class="p">(</span><span class="n">basecall_group</span><span class="p">)</span>
            <span class="n">bc_subgrp</span> <span class="o">=</span> <span class="n">bc_grp</span><span class="o">.</span><span class="n">create_group</span><span class="p">(</span><span class="n">basecall_subgroup</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span>
                <span class="n">basecall_group</span> <span class="o">+</span> <span class="s1">&#39; exists and --overwrite is not set.&#39;</span><span class="p">)</span>
</div>
    <span class="k">return</span> <span class="n">read_id</span>

<div class="viewcode-block" id="get_read_ids_and_prep_fastq_slot"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.get_read_ids_and_prep_fastq_slot">[docs]</a><span class="k">def</span> <span class="nf">get_read_ids_and_prep_fastq_slot</span><span class="p">(</span>
        <span class="n">fast5s_dir</span><span class="p">,</span> <span class="n">basecall_group</span><span class="p">,</span> <span class="n">basecall_subgroup</span><span class="p">,</span> <span class="n">overwrite</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Extract read id from /Raw group and prep fastq slots for annotation with</span>
<span class="sd">    associated FASTQ files.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">VERBOSE</span><span class="p">:</span> <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
            <span class="s1">&#39;Preparing reads and extracting read identifiers.</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="n">been_warned_overwrite</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="n">been_warned_unique</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="n">fast5s_read_ids</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="c1"># walk through directory structure searching for fast5 files</span>
    <span class="k">for</span> <span class="n">root</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">fns</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">walk</span><span class="p">(</span><span class="n">fast5s_dir</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">fn</span> <span class="ow">in</span> <span class="n">fnmatch</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">fns</span><span class="p">,</span> <span class="s1">&#39;*.fast5&#39;</span><span class="p">):</span>
            <span class="n">fast5_fn</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">root</span><span class="p">,</span> <span class="n">fn</span><span class="p">)</span>
            <span class="k">with</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="n">fast5_fn</span><span class="p">)</span> <span class="k">as</span> <span class="n">fast5_data</span><span class="p">:</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">read_id</span> <span class="o">=</span> <span class="n">prep_fast5_for_fastq</span><span class="p">(</span>
                        <span class="n">fast5_data</span><span class="p">,</span> <span class="n">basecall_group</span><span class="p">,</span> <span class="n">basecall_subgroup</span><span class="p">,</span> <span class="n">overwrite</span><span class="p">)</span>
                <span class="k">except</span> <span class="ne">NotImplementedError</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">VERBOSE</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">been_warned_overwrite</span><span class="p">:</span>
                        <span class="n">been_warned_overwrite</span> <span class="o">=</span> <span class="kc">True</span>
                        <span class="n">_warning_message</span><span class="p">(</span>
                            <span class="s1">&#39;Basecalls exsit in &#39;</span> <span class="o">+</span> <span class="n">basecall_group</span> <span class="o">+</span> <span class="s1">&#39; slot. &#39;</span> <span class="o">+</span>
                            <span class="s1">&#39;Set --overwrite option to overwrite these &#39;</span> <span class="o">+</span>
                            <span class="s1">&#39;basecalls in this slot.&#39;</span><span class="p">)</span>
                    <span class="k">continue</span>
            <span class="k">if</span> <span class="n">read_id</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="k">if</span> <span class="n">read_id</span> <span class="ow">in</span> <span class="n">fast5s_read_ids</span><span class="p">:</span>
                <span class="c1"># Warn non-unique read_ids in directory</span>
                <span class="k">if</span> <span class="n">VERBOSE</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">been_warned_unique</span><span class="p">:</span>
                    <span class="n">been_warned_unique</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="n">_warning_message</span><span class="p">(</span>
                        <span class="s1">&#39;Multiple FAST5 files contain the same read &#39;</span> <span class="o">+</span>
                        <span class="s1">&#39;identifiers. Ensure that FAST5 files are from &#39;</span> <span class="o">+</span>
                        <span class="s1">&#39;a single run.&#39;</span><span class="p">)</span>
                <span class="k">continue</span>

            <span class="n">fast5s_read_ids</span><span class="p">[</span><span class="n">read_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">fast5_fn</span>
</div>
    <span class="k">return</span> <span class="n">fast5s_read_ids</span>

<div class="viewcode-block" id="parse_fast5s_wo_index"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.parse_fast5s_wo_index">[docs]</a><span class="k">def</span> <span class="nf">parse_fast5s_wo_index</span><span class="p">(</span>
        <span class="n">fast5_basedirs</span><span class="p">,</span> <span class="n">corrected_group</span><span class="p">,</span> <span class="n">basecall_subgroups</span><span class="p">,</span> <span class="n">rna</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Parse re-squiggled reads data from a list of fast5 directories</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">files</span> <span class="o">=</span> <span class="p">[</span><span class="n">fn</span> <span class="k">for</span> <span class="n">fast5_basedir</span> <span class="ow">in</span> <span class="n">fast5_basedirs</span>
             <span class="k">for</span> <span class="n">fn</span> <span class="ow">in</span> <span class="n">get_files_list</span><span class="p">(</span><span class="n">fast5_basedir</span><span class="p">)]</span>
    <span class="n">raw_read_coverage</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">read_fn</span> <span class="ow">in</span> <span class="n">files</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">read_data</span> <span class="o">=</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="n">read_fn</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
            <span class="c1"># probably truncated file</span>
            <span class="k">continue</span>
        <span class="k">for</span> <span class="n">basecall_subgroup</span> <span class="ow">in</span> <span class="n">basecall_subgroups</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">corr_data</span> <span class="o">=</span> <span class="n">read_data</span><span class="p">[</span><span class="s1">&#39;/&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">((</span>
                    <span class="s1">&#39;/Analyses&#39;</span><span class="p">,</span> <span class="n">corrected_group</span><span class="p">,</span> <span class="n">basecall_subgroup</span><span class="p">))]</span>
            <span class="k">except</span><span class="p">:</span>
                <span class="k">continue</span>

            <span class="k">try</span><span class="p">:</span>
                <span class="n">align_data</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">corr_data</span><span class="p">[</span><span class="s1">&#39;Alignment&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">attrs</span><span class="o">.</span><span class="n">items</span><span class="p">())</span>
                <span class="n">read_start_rel_to_raw</span> <span class="o">=</span> <span class="n">corr_data</span><span class="p">[</span><span class="s1">&#39;Events&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span>
                    <span class="s1">&#39;read_start_rel_to_raw&#39;</span><span class="p">]</span>
            <span class="k">except</span><span class="p">:</span>
                <span class="c1"># don&#39;t warn here since errored out reads will have get here, but</span>
                <span class="c1"># not have alignment and events stored, so just skip these reads</span>
                <span class="k">continue</span>
            <span class="n">chrm</span> <span class="o">=</span> <span class="n">align_data</span><span class="p">[</span><span class="s1">&#39;mapped_chrom&#39;</span><span class="p">]</span>
            <span class="n">strand</span> <span class="o">=</span> <span class="n">align_data</span><span class="p">[</span><span class="s1">&#39;mapped_strand&#39;</span><span class="p">]</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">chrm</span> <span class="o">=</span> <span class="n">chrm</span><span class="o">.</span><span class="n">decode</span><span class="p">()</span>
                <span class="n">strand</span> <span class="o">=</span> <span class="n">strand</span><span class="o">.</span><span class="n">decode</span><span class="p">()</span>
            <span class="k">except</span><span class="p">:</span>
                <span class="k">pass</span>
            <span class="n">raw_read_coverage</span><span class="p">[(</span><span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">)]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="n">readData</span><span class="p">(</span>
                    <span class="n">align_data</span><span class="p">[</span><span class="s1">&#39;mapped_start&#39;</span><span class="p">],</span> <span class="n">align_data</span><span class="p">[</span><span class="s1">&#39;mapped_end&#39;</span><span class="p">],</span>
                    <span class="kc">False</span><span class="p">,</span> <span class="n">read_start_rel_to_raw</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">read_fn</span><span class="p">,</span>
                    <span class="n">corrected_group</span> <span class="o">+</span> <span class="s1">&#39;/&#39;</span> <span class="o">+</span> <span class="n">basecall_subgroup</span><span class="p">,</span> <span class="n">rna</span><span class="p">))</span>

        <span class="n">read_data</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</div>
    <span class="k">return</span> <span class="nb">dict</span><span class="p">(</span><span class="n">raw_read_coverage</span><span class="p">)</span>

<div class="viewcode-block" id="convert_index"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.convert_index">[docs]</a><span class="k">def</span> <span class="nf">convert_index</span><span class="p">(</span><span class="n">index_data</span><span class="p">,</span> <span class="n">fast5s_dir</span><span class="p">,</span> <span class="n">corr_grp</span><span class="p">,</span> <span class="n">new_corr_grp</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Convert an index and save under a new corrected group. Mostly for</span>
<span class="sd">    model_resquiggle</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">new_index_data</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">),</span> <span class="n">cs_raw_data</span> <span class="ow">in</span> <span class="n">index_data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">for</span> <span class="p">(</span><span class="n">from_base_fn</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">rsrtr</span><span class="p">,</span> <span class="n">c_grp</span><span class="p">,</span> <span class="n">s_grp</span><span class="p">,</span>
             <span class="n">filtered</span><span class="p">,</span> <span class="n">rna</span><span class="p">)</span> <span class="ow">in</span> <span class="n">cs_raw_data</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">c_grp</span> <span class="o">!=</span> <span class="n">corr_grp</span><span class="p">:</span> <span class="k">continue</span>
            <span class="n">new_index_data</span><span class="o">.</span><span class="n">append</span><span class="p">(((</span><span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">),</span> <span class="p">(</span>
                <span class="n">from_base_fn</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">rsrtr</span><span class="p">,</span>
                <span class="n">new_corr_grp</span><span class="p">,</span> <span class="n">s_grp</span><span class="p">,</span> <span class="n">filtered</span><span class="p">,</span> <span class="n">rna</span><span class="p">)))</span>

    <span class="n">fast5s_dir</span> <span class="o">=</span> <span class="p">(</span><span class="n">fast5s_dir</span> <span class="k">if</span> <span class="n">fast5s_dir</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)</span> <span class="k">else</span>
                  <span class="n">fast5s_dir</span> <span class="o">+</span> <span class="s1">&#39;/&#39;</span><span class="p">)</span>
    <span class="n">new_index_fn</span> <span class="o">=</span> <span class="n">get_index_fn</span><span class="p">(</span><span class="n">fast5s_dir</span><span class="p">,</span> <span class="n">new_corr_grp</span><span class="p">)</span>
    <span class="n">write_index_file</span><span class="p">(</span><span class="n">new_index_data</span><span class="p">,</span> <span class="n">new_index_fn</span><span class="p">,</span> <span class="n">fast5s_dir</span><span class="p">)</span>
</div>
    <span class="k">return</span>

<div class="viewcode-block" id="parse_fast5s_w_index"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.parse_fast5s_w_index">[docs]</a><span class="k">def</span> <span class="nf">parse_fast5s_w_index</span><span class="p">(</span><span class="n">fast5s_dir</span><span class="p">,</span> <span class="n">corr_grp</span><span class="p">,</span> <span class="n">subgroups</span><span class="p">,</span> <span class="n">new_corr_grp</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Use index file to parse information about a set of reads</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">fast5s_dir</span> <span class="o">=</span> <span class="p">(</span><span class="n">fast5s_dir</span> <span class="k">if</span> <span class="n">fast5s_dir</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)</span> <span class="k">else</span>
                  <span class="n">fast5s_dir</span> <span class="o">+</span> <span class="s1">&#39;/&#39;</span><span class="p">)</span>
    <span class="n">index_fn</span> <span class="o">=</span> <span class="n">get_index_fn</span><span class="p">(</span><span class="n">fast5s_dir</span><span class="p">,</span> <span class="n">corr_grp</span><span class="p">)</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="kn">import</span> <span class="nn">cPickle</span> <span class="k">as</span> <span class="nn">pickle</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="kn">import</span> <span class="nn">pickle</span>
    <span class="k">with</span> <span class="n">io</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">index_fn</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">index_fp</span><span class="p">:</span>
        <span class="n">index_data</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">index_fp</span><span class="p">)</span>
    <span class="n">raw_read_coverage</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">),</span> <span class="n">cs_raw_data</span> <span class="ow">in</span> <span class="n">index_data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">cs_data</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">readData</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">filtered</span><span class="p">,</span> <span class="n">rsrtr</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span>
                     <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">fast5s_dir</span><span class="p">,</span> <span class="n">from_base_fn</span><span class="p">),</span>
                     <span class="n">corr_grp</span> <span class="o">+</span> <span class="s1">&#39;/&#39;</span> <span class="o">+</span> <span class="n">s_grp</span><span class="p">,</span> <span class="n">rna</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">from_base_fn</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">rsrtr</span><span class="p">,</span> <span class="n">c_grp</span><span class="p">,</span>
            <span class="n">s_grp</span><span class="p">,</span> <span class="n">filtered</span><span class="p">,</span> <span class="n">rna</span> <span class="ow">in</span> <span class="n">cs_raw_data</span>
            <span class="k">if</span> <span class="n">c_grp</span> <span class="o">==</span> <span class="n">corr_grp</span> <span class="ow">and</span> <span class="n">s_grp</span> <span class="ow">in</span> <span class="n">subgroups</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">filtered</span><span class="p">]</span>
        <span class="n">raw_read_coverage</span><span class="p">[(</span><span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">)]</span> <span class="o">=</span> <span class="n">cs_data</span>
    <span class="k">if</span> <span class="n">new_corr_grp</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="c1"># convert corrected group to new corrected group for</span>
        <span class="c1"># model re-squiggle</span>
        <span class="n">convert_index</span><span class="p">(</span><span class="n">index_data</span><span class="p">,</span> <span class="n">fast5s_dir</span><span class="p">,</span> <span class="n">corr_grp</span><span class="p">,</span> <span class="n">new_corr_grp</span><span class="p">)</span>
</div>
    <span class="k">return</span> <span class="n">raw_read_coverage</span>

<div class="viewcode-block" id="merge_cov"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.merge_cov">[docs]</a><span class="k">def</span> <span class="nf">merge_cov</span><span class="p">(</span><span class="n">w_index_covs</span><span class="p">,</span> <span class="n">wo_index_cov</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Merge coverage from serveral parsed sets of data</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">all_covs</span> <span class="o">=</span> <span class="n">w_index_covs</span> <span class="o">+</span> <span class="p">[</span><span class="n">wo_index_cov</span><span class="p">,]</span>
    <span class="n">raw_read_coverage</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">chrm_strand</span> <span class="ow">in</span> <span class="nb">set</span><span class="p">([</span><span class="n">cs</span> <span class="k">for</span> <span class="n">d_cov</span> <span class="ow">in</span> <span class="n">all_covs</span> <span class="k">for</span> <span class="n">cs</span> <span class="ow">in</span> <span class="n">d_cov</span><span class="p">]):</span>
        <span class="k">for</span> <span class="n">dir_cov</span> <span class="ow">in</span> <span class="n">all_covs</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">chrm_strand</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">dir_cov</span><span class="p">:</span> <span class="k">continue</span>
            <span class="n">raw_read_coverage</span><span class="p">[</span><span class="n">chrm_strand</span><span class="p">]</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">dir_cov</span><span class="p">[</span><span class="n">chrm_strand</span><span class="p">])</span>
</div>
    <span class="k">return</span> <span class="nb">dict</span><span class="p">(</span><span class="n">raw_read_coverage</span><span class="p">)</span>

<div class="viewcode-block" id="parse_fast5s"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.parse_fast5s">[docs]</a><span class="k">def</span> <span class="nf">parse_fast5s</span><span class="p">(</span><span class="n">fast5_basedirs</span><span class="p">,</span> <span class="n">corrected_group</span><span class="p">,</span> <span class="n">basecall_subgroups</span><span class="p">,</span>
                 <span class="n">new_corr_grp</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">rna</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Parse data from a list of re-squiggle fast5 directories</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">VERBOSE</span><span class="p">:</span> <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;Parsing tombo index file(s).</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="n">wo_index_dirs</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">w_index_covs</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">warn_index</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="c1"># determine if index exists for each directory and load appropriately</span>
    <span class="k">for</span> <span class="n">fast5s_dir</span> <span class="ow">in</span> <span class="n">fast5_basedirs</span><span class="p">:</span>
        <span class="n">fast5s_dir</span> <span class="o">=</span> <span class="p">(</span><span class="n">fast5s_dir</span> <span class="k">if</span> <span class="n">fast5s_dir</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)</span> <span class="k">else</span>
                      <span class="n">fast5s_dir</span> <span class="o">+</span> <span class="s1">&#39;/&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">get_index_fn</span><span class="p">(</span><span class="n">fast5s_dir</span><span class="p">,</span> <span class="n">corrected_group</span><span class="p">)):</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">w_index_covs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">parse_fast5s_w_index</span><span class="p">(</span>
                    <span class="n">fast5s_dir</span><span class="p">,</span> <span class="n">corrected_group</span><span class="p">,</span> <span class="n">basecall_subgroups</span><span class="p">,</span>
                    <span class="n">new_corr_grp</span><span class="p">))</span>
            <span class="k">except</span><span class="p">:</span>
                <span class="k">raise</span>
                <span class="n">_warning_message</span><span class="p">(</span>
                    <span class="s1">&#39;Failed to parse tombo index file for &#39;</span> <span class="o">+</span>
                    <span class="n">fast5s_dir</span> <span class="o">+</span> <span class="s1">&#39; directory.&#39;</span><span class="p">)</span>
                <span class="n">wo_index_dirs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fast5s_dir</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">warn_index</span><span class="p">:</span>
                <span class="n">_warning_message</span><span class="p">(</span>
                    <span class="s1">&#39;Tombo index file does not exist for one or more &#39;</span> <span class="o">+</span>
                    <span class="s1">&#39;directories. If --skip-index was not set for &#39;</span> <span class="o">+</span>
                    <span class="s1">&#39;re-squiggle command, ensure that the specified &#39;</span> <span class="o">+</span>
                    <span class="s1">&#39;directory is the same as for the re-squiggle command.</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
                <span class="n">warn_index</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="n">wo_index_dirs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fast5s_dir</span><span class="p">)</span>
    <span class="n">wo_index_cov</span> <span class="o">=</span> <span class="n">parse_fast5s_wo_index</span><span class="p">(</span>
        <span class="n">wo_index_dirs</span><span class="p">,</span> <span class="n">corrected_group</span><span class="p">,</span> <span class="n">basecall_subgroups</span><span class="p">,</span> <span class="n">rna</span><span class="p">)</span>
    <span class="n">raw_read_coverage</span> <span class="o">=</span> <span class="n">merge_cov</span><span class="p">(</span><span class="n">w_index_covs</span><span class="p">,</span> <span class="n">wo_index_cov</span><span class="p">)</span>
</div>
    <span class="k">return</span> <span class="n">raw_read_coverage</span>


<span class="c1">###################################</span>
<span class="c1">###### Statistical Functions ######</span>
<span class="c1">###################################</span>

<span class="c1"># Some of these functions should likely be re-factored to tombo_stats</span>

<div class="viewcode-block" id="parse_pore_model"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.parse_pore_model">[docs]</a><span class="k">def</span> <span class="nf">parse_pore_model</span><span class="p">(</span><span class="n">pore_model_fn</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Parse pore model for pA normalization (Deprecated)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">pore_model</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;mean&#39;</span><span class="p">:{},</span> <span class="s1">&#39;inv_var&#39;</span><span class="p">:{}}</span>
    <span class="k">with</span> <span class="n">io</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">pore_model_fn</span><span class="p">)</span> <span class="k">as</span> <span class="n">fp</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">fp</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;#&#39;</span><span class="p">):</span> <span class="k">continue</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">kmer</span><span class="p">,</span> <span class="n">lev_mean</span><span class="p">,</span> <span class="n">lev_stdev</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()[:</span><span class="mi">3</span><span class="p">]</span>
                <span class="n">lev_mean</span><span class="p">,</span> <span class="n">lev_stdev</span> <span class="o">=</span> <span class="nb">map</span><span class="p">(</span><span class="nb">float</span><span class="p">,</span> <span class="p">(</span><span class="n">lev_mean</span><span class="p">,</span> <span class="n">lev_stdev</span><span class="p">))</span>
            <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
                <span class="c1"># header or other non-kmer field</span>
                <span class="k">continue</span>
            <span class="n">pore_model</span><span class="p">[</span><span class="s1">&#39;mean&#39;</span><span class="p">][</span><span class="n">kmer</span><span class="p">]</span> <span class="o">=</span> <span class="n">lev_mean</span>
            <span class="n">pore_model</span><span class="p">[</span><span class="s1">&#39;inv_var&#39;</span><span class="p">][</span><span class="n">kmer</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">/</span> <span class="p">(</span><span class="n">lev_stdev</span> <span class="o">*</span> <span class="n">lev_stdev</span><span class="p">)</span>
</div>
    <span class="k">return</span> <span class="n">pore_model</span>

<div class="viewcode-block" id="calc_kmer_fitted_shift_scale"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.calc_kmer_fitted_shift_scale">[docs]</a><span class="k">def</span> <span class="nf">calc_kmer_fitted_shift_scale</span><span class="p">(</span><span class="n">pore_model</span><span class="p">,</span> <span class="n">events_means</span><span class="p">,</span> <span class="n">events_kmers</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Compute fitted shift and scale parameters for pA normalization</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">r_model_means</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">pore_model</span><span class="p">[</span><span class="s1">&#39;mean&#39;</span><span class="p">][</span><span class="n">kmer</span><span class="p">]</span>
                              <span class="k">for</span> <span class="n">kmer</span> <span class="ow">in</span> <span class="n">events_kmers</span><span class="p">])</span>
    <span class="n">r_model_inv_vars</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">pore_model</span><span class="p">[</span><span class="s1">&#39;inv_var&#39;</span><span class="p">][</span><span class="n">kmer</span><span class="p">]</span>
                                 <span class="k">for</span> <span class="n">kmer</span> <span class="ow">in</span> <span class="n">events_kmers</span><span class="p">])</span>
    <span class="n">model_mean_var</span> <span class="o">=</span> <span class="n">r_model_means</span> <span class="o">*</span> <span class="n">r_model_inv_vars</span>
    <span class="c1"># prep kmer model coefficient matrix for the k-mers from this read</span>
    <span class="n">model_mean_var_sum</span> <span class="o">=</span> <span class="n">model_mean_var</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
    <span class="n">coef_mat</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">((</span>
        <span class="p">(</span><span class="n">r_model_inv_vars</span><span class="o">.</span><span class="n">sum</span><span class="p">(),</span> <span class="n">model_mean_var_sum</span><span class="p">),</span>
        <span class="p">(</span><span class="n">model_mean_var_sum</span><span class="p">,</span> <span class="p">(</span><span class="n">model_mean_var</span> <span class="o">*</span> <span class="n">r_model_means</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">())))</span>

    <span class="c1"># prep dependent values from this reads true events</span>
    <span class="n">r_event_var</span> <span class="o">=</span> <span class="n">events_means</span> <span class="o">*</span> <span class="n">r_model_inv_vars</span>
    <span class="n">r_event_var_mean</span> <span class="o">=</span> <span class="n">r_event_var</span> <span class="o">*</span> <span class="n">r_model_means</span>
    <span class="n">dep_vect</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">((</span><span class="n">r_event_var</span><span class="o">.</span><span class="n">sum</span><span class="p">(),</span> <span class="n">r_event_var_mean</span><span class="o">.</span><span class="n">sum</span><span class="p">()))</span>

    <span class="n">shift</span><span class="p">,</span> <span class="n">scale</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">solve</span><span class="p">(</span><span class="n">coef_mat</span><span class="p">,</span> <span class="n">dep_vect</span><span class="p">)</span>
</div>
    <span class="k">return</span> <span class="n">shift</span><span class="p">,</span> <span class="n">scale</span>

<div class="viewcode-block" id="get_valid_cpts"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.get_valid_cpts">[docs]</a><span class="k">def</span> <span class="nf">get_valid_cpts</span><span class="p">(</span><span class="n">norm_signal</span><span class="p">,</span> <span class="n">running_stat_width</span><span class="p">,</span> <span class="n">num_events</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get valid changepoints given largest differences in neighboring</span>
<span class="sd">    moving windows</span>

<span class="sd">    Note that this method is completely vectorized, but allows segments</span>
<span class="sd">    as small as 2 observations. This should be okay R9+, but is problematic</span>
<span class="sd">    for &lt;=R7 and RNA</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">raw_cumsum</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">cumsum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([[</span><span class="mi">0</span><span class="p">],</span> <span class="n">norm_signal</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]]))</span>
    <span class="c1"># get difference between all neighboring running_stat_width regions</span>
    <span class="n">running_diffs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span>
        <span class="p">(</span><span class="mi">2</span> <span class="o">*</span> <span class="n">raw_cumsum</span><span class="p">[</span><span class="n">running_stat_width</span><span class="p">:</span><span class="o">-</span><span class="n">running_stat_width</span><span class="p">])</span> <span class="o">-</span>
        <span class="n">raw_cumsum</span><span class="p">[:</span><span class="o">-</span><span class="mi">2</span><span class="o">*</span><span class="n">running_stat_width</span><span class="p">]</span> <span class="o">-</span>
        <span class="n">raw_cumsum</span><span class="p">[</span><span class="mi">2</span><span class="o">*</span><span class="n">running_stat_width</span><span class="p">:])</span>
    <span class="n">not_peaks</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">logical_not</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">logical_and</span><span class="p">(</span>
        <span class="n">running_diffs</span> <span class="o">&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([[</span><span class="mi">0</span><span class="p">],</span> <span class="n">running_diffs</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]]),</span>
        <span class="n">running_diffs</span> <span class="o">&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">running_diffs</span><span class="p">[</span><span class="mi">1</span><span class="p">:],</span> <span class="p">[</span><span class="mi">0</span><span class="p">]])))</span>
    <span class="n">running_diffs</span><span class="p">[</span><span class="n">not_peaks</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">valid_cpts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span>
        <span class="n">running_diffs</span><span class="p">)[::</span><span class="o">-</span><span class="mi">1</span><span class="p">][:</span><span class="n">num_events</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">int64</span><span class="p">)</span> <span class="o">+</span> <span class="n">running_stat_width</span>
</div>
    <span class="k">return</span> <span class="n">valid_cpts</span>

<div class="viewcode-block" id="estimate_global_scale"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.estimate_global_scale">[docs]</a><span class="k">def</span> <span class="nf">estimate_global_scale</span><span class="p">(</span><span class="n">fast5_fns</span><span class="p">,</span> <span class="n">num_reads</span><span class="o">=</span><span class="n">NUM_READS_FOR_SCALE</span><span class="p">):</span>
    <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;Estimating global scale parameter</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">fast5_fns</span><span class="p">)</span>
    <span class="n">read_mads</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">fast5_fn</span> <span class="ow">in</span> <span class="n">fast5_fns</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">with</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="n">fast5_fn</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fast5_data</span><span class="p">:</span>
                <span class="n">all_sig</span> <span class="o">=</span> <span class="n">get_raw_read_slot</span><span class="p">(</span><span class="n">fast5_data</span><span class="p">)[</span><span class="s1">&#39;Signal&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value</span>
            <span class="n">shift</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">all_sig</span><span class="p">)</span>
            <span class="n">read_mads</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">all_sig</span> <span class="o">-</span> <span class="n">shift</span><span class="p">)))</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">read_mads</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">num_reads</span><span class="p">:</span>
            <span class="k">break</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">read_mads</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">_error_message_and_exit</span><span class="p">(</span>
            <span class="s1">&#39;No reads contain raw signal for &#39;</span> <span class="o">+</span>
            <span class="s1">&#39;global scale parameter estimation.&#39;</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">read_mads</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">num_reads</span><span class="p">:</span>
        <span class="n">_warning_message</span><span class="p">(</span>
            <span class="s1">&#39;Few reads contain raw signal for global scale parameter &#39;</span> <span class="o">+</span>
            <span class="s1">&#39;estimation. Results may not be optimal.&#39;</span><span class="p">)</span>
</div>
    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">read_mads</span><span class="p">)</span>

<div class="viewcode-block" id="normalize_raw_signal"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.normalize_raw_signal">[docs]</a><span class="k">def</span> <span class="nf">normalize_raw_signal</span><span class="p">(</span>
        <span class="n">all_raw_signal</span><span class="p">,</span> <span class="n">read_start_rel_to_raw</span><span class="p">,</span> <span class="n">read_obs_len</span><span class="p">,</span>
        <span class="n">norm_type</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">channel_info</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">outlier_thresh</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">shift</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">scale</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">lower_lim</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">upper_lim</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">pore_model</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">event_means</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">event_kmers</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">const_scale</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Apply scaling and windsorizing parameters to normalize raw signal</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">norm_type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">NORM_TYPES</span> <span class="ow">and</span> <span class="p">(</span><span class="n">shift</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">scale</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span>
            <span class="s1">&#39;Normalization type &#39;</span> <span class="o">+</span> <span class="n">norm_type</span> <span class="o">+</span> <span class="s1">&#39; is not a valid &#39;</span> <span class="o">+</span>
            <span class="s1">&#39;option and shift or scale parameters were not provided.&#39;</span><span class="p">)</span>

    <span class="n">raw_signal</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">all_raw_signal</span><span class="p">[</span>
        <span class="n">read_start_rel_to_raw</span><span class="p">:</span>
        <span class="n">read_start_rel_to_raw</span> <span class="o">+</span> <span class="n">read_obs_len</span><span class="p">])</span>
    <span class="k">if</span> <span class="n">shift</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">scale</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">norm_type</span> <span class="o">==</span> <span class="s1">&#39;none&#39;</span><span class="p">:</span>
            <span class="n">shift</span><span class="p">,</span> <span class="n">scale</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span>
        <span class="k">elif</span> <span class="n">norm_type</span> <span class="ow">in</span> <span class="p">(</span><span class="s1">&#39;pA_raw&#39;</span><span class="p">,</span> <span class="s1">&#39;pA&#39;</span><span class="p">):</span>
            <span class="c1"># correct raw signal as described here:</span>
            <span class="c1">#     https://community.nanoporetech.com</span>
            <span class="c1">#           /posts/squiggle-plot-for-raw-data</span>
            <span class="n">shift</span><span class="p">,</span> <span class="n">scale</span> <span class="o">=</span> <span class="p">(</span>
                <span class="o">-</span><span class="mi">1</span> <span class="o">*</span> <span class="n">channel_info</span><span class="o">.</span><span class="n">offset</span><span class="p">,</span>
                <span class="n">channel_info</span><span class="o">.</span><span class="n">digitisation</span> <span class="o">/</span> <span class="n">channel_info</span><span class="o">.</span><span class="n">range</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">norm_type</span> <span class="o">==</span> <span class="s1">&#39;pA&#39;</span><span class="p">:</span>
                <span class="c1"># perform k-mer model fitted correction as in</span>
                <span class="c1"># nanocorr/nanopolish/ONT</span>
                <span class="n">fit_shift</span><span class="p">,</span> <span class="n">fit_scale</span> <span class="o">=</span> <span class="n">calc_kmer_fitted_shift_scale</span><span class="p">(</span>
                    <span class="n">pore_model</span><span class="p">,</span> <span class="n">event_means</span><span class="p">,</span> <span class="n">event_kmers</span><span class="p">)</span>
                <span class="c1"># apply shift and scale values fitted from kmer</span>
                <span class="c1"># conditional model after raw DAC scaling</span>
                <span class="n">shift</span> <span class="o">=</span> <span class="n">shift</span> <span class="o">+</span> <span class="p">(</span><span class="n">fit_shift</span> <span class="o">*</span> <span class="n">scale</span><span class="p">)</span>
                <span class="n">scale</span> <span class="o">=</span> <span class="n">scale</span> <span class="o">*</span> <span class="n">fit_scale</span>
        <span class="k">elif</span> <span class="n">norm_type</span> <span class="o">==</span> <span class="s1">&#39;median&#39;</span><span class="p">:</span>
            <span class="n">shift</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">raw_signal</span><span class="p">)</span>
            <span class="n">scale</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">raw_signal</span> <span class="o">-</span> <span class="n">shift</span><span class="p">))</span>
        <span class="k">elif</span> <span class="n">norm_type</span> <span class="o">==</span> <span class="s1">&#39;median_const_scale&#39;</span><span class="p">:</span>
            <span class="k">assert</span> <span class="n">const_scale</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
            <span class="n">shift</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">raw_signal</span><span class="p">)</span>
            <span class="n">scale</span> <span class="o">=</span> <span class="n">const_scale</span>
        <span class="k">elif</span> <span class="n">norm_type</span> <span class="o">==</span> <span class="s1">&#39;robust_median&#39;</span><span class="p">:</span>
            <span class="n">shift</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">raw_signal</span><span class="p">,</span> <span class="n">ROBUST_QUANTS</span><span class="p">))</span>
            <span class="n">scale</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">raw_signal</span> <span class="o">-</span> <span class="n">read_robust_med</span><span class="p">))</span>

    <span class="n">raw_signal</span> <span class="o">=</span> <span class="p">(</span><span class="n">raw_signal</span> <span class="o">-</span> <span class="n">shift</span><span class="p">)</span> <span class="o">/</span> <span class="n">scale</span>

    <span class="k">if</span> <span class="n">outlier_thresh</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">or</span> <span class="p">(</span>
            <span class="n">lower_lim</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">upper_lim</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">outlier_thresh</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">read_med</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">raw_signal</span><span class="p">)</span>
            <span class="n">read_mad</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">raw_signal</span> <span class="o">-</span> <span class="n">read_med</span><span class="p">))</span>
            <span class="n">lower_lim</span> <span class="o">=</span> <span class="n">read_med</span> <span class="o">-</span> <span class="p">(</span><span class="n">read_mad</span> <span class="o">*</span> <span class="n">outlier_thresh</span><span class="p">)</span>
            <span class="n">upper_lim</span> <span class="o">=</span> <span class="n">read_med</span> <span class="o">+</span> <span class="p">(</span><span class="n">read_mad</span> <span class="o">*</span> <span class="n">outlier_thresh</span><span class="p">)</span>
        <span class="n">raw_signal</span> <span class="o">=</span> <span class="n">c_apply_outlier_thresh</span><span class="p">(</span><span class="n">raw_signal</span><span class="p">,</span> <span class="n">lower_lim</span><span class="p">,</span> <span class="n">upper_lim</span><span class="p">)</span>
</div>
    <span class="k">return</span> <span class="n">raw_signal</span><span class="p">,</span> <span class="n">scaleValues</span><span class="p">(</span><span class="n">shift</span><span class="p">,</span> <span class="n">scale</span><span class="p">,</span> <span class="n">lower_lim</span><span class="p">,</span> <span class="n">upper_lim</span><span class="p">)</span>


<span class="c1">###########################################</span>
<span class="c1">###### Events Table Access Functions ######</span>
<span class="c1">###########################################</span>

<div class="viewcode-block" id="get_multiple_slots_read_centric"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.get_multiple_slots_read_centric">[docs]</a><span class="k">def</span> <span class="nf">get_multiple_slots_read_centric</span><span class="p">(</span><span class="n">r_data</span><span class="p">,</span> <span class="n">slot_names</span><span class="p">,</span> <span class="n">corr_grp</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Extract read-centric slot_names from this read&#39;s Events table</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">do_close</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">r_data</span><span class="p">,</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">):</span>
            <span class="n">do_close</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="n">corr_grp</span> <span class="o">=</span> <span class="n">r_data</span><span class="o">.</span><span class="n">corr_group</span>
            <span class="n">r_data</span> <span class="o">=</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="n">r_data</span><span class="o">.</span><span class="n">fn</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span>
        <span class="n">event_slot_name</span> <span class="o">=</span> <span class="s1">&#39;/&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">((</span><span class="s1">&#39;/Analyses&#39;</span><span class="p">,</span> <span class="n">corr_grp</span><span class="p">,</span> <span class="s1">&#39;Events&#39;</span><span class="p">))</span>
        <span class="c1"># note that it&#39;s more efficient to try to access the slot</span>
        <span class="c1"># and except the error that check if the slot exists first</span>
        <span class="n">r_event_data</span> <span class="o">=</span> <span class="n">r_data</span><span class="p">[</span><span class="n">event_slot_name</span><span class="p">]</span><span class="o">.</span><span class="n">value</span>
        <span class="k">if</span> <span class="n">do_close</span><span class="p">:</span> <span class="n">r_data</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="c1"># probably truncated file or events don&#39;t exist</span>
        <span class="k">return</span> <span class="p">[</span><span class="kc">None</span><span class="p">,]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">slot_names</span><span class="p">)</span>
</div>
    <span class="k">return</span> <span class="p">[</span><span class="n">r_event_data</span><span class="p">[</span><span class="n">slot_name</span><span class="p">]</span> <span class="k">for</span> <span class="n">slot_name</span> <span class="ow">in</span> <span class="n">slot_names</span><span class="p">]</span>

<div class="viewcode-block" id="get_single_slot_read_centric"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.get_single_slot_read_centric">[docs]</a><span class="k">def</span> <span class="nf">get_single_slot_read_centric</span><span class="p">(</span><span class="n">r_data</span><span class="p">,</span> <span class="n">slot_name</span><span class="p">,</span> <span class="n">corr_grp</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Extract read-centric slot_name from this read&#39;s Events table</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="c1"># if r_data is an open h5py object then don&#39;t open the filename</span>
        <span class="n">do_close</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">r_data</span><span class="p">,</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">):</span>
            <span class="n">do_close</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="n">corr_grp</span> <span class="o">=</span> <span class="n">r_data</span><span class="o">.</span><span class="n">corr_group</span>
            <span class="n">r_data</span> <span class="o">=</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="n">r_data</span><span class="o">.</span><span class="n">fn</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span>
        <span class="c1"># note that it&#39;s more efficient to try to access the slot</span>
        <span class="c1"># and except the error that check if the slot exists first</span>
        <span class="n">r_slot_values</span> <span class="o">=</span> <span class="n">r_data</span><span class="p">[</span><span class="s1">&#39;/&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">((</span><span class="s1">&#39;/Analyses&#39;</span><span class="p">,</span> <span class="n">corr_grp</span><span class="p">,</span> <span class="s1">&#39;Events&#39;</span><span class="p">))][</span>
            <span class="n">slot_name</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">do_close</span><span class="p">:</span> <span class="n">r_data</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="c1"># probably truncated file or events don&#39;t exist</span>
        <span class="k">return</span> <span class="kc">None</span>
</div>
    <span class="k">return</span> <span class="n">r_slot_values</span>

<div class="viewcode-block" id="get_single_slot_genome_centric"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.get_single_slot_genome_centric">[docs]</a><span class="k">def</span> <span class="nf">get_single_slot_genome_centric</span><span class="p">(</span><span class="n">r_data</span><span class="p">,</span> <span class="n">slot_name</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Extract genome-centric slot_name from this read&#39;s Events table</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">r_slot_values</span> <span class="o">=</span> <span class="n">get_single_slot_read_centric</span><span class="p">(</span><span class="n">r_data</span><span class="p">,</span> <span class="n">slot_name</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">r_slot_values</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="k">if</span> <span class="n">r_data</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;-&#39;</span><span class="p">:</span>
        <span class="n">r_slot_values</span> <span class="o">=</span> <span class="n">r_slot_values</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
</div>
    <span class="k">return</span> <span class="n">r_slot_values</span>

<div class="viewcode-block" id="get_mean_slot_genome_centric"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.get_mean_slot_genome_centric">[docs]</a><span class="k">def</span> <span class="nf">get_mean_slot_genome_centric</span><span class="p">(</span><span class="n">cs_reads</span><span class="p">,</span> <span class="n">chrm_len</span><span class="p">,</span> <span class="n">slot_name</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get the mean over all reads at each covered genomic location for this</span>
<span class="sd">    slots value</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">base_sums</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">chrm_len</span><span class="p">)</span>
    <span class="n">base_cov</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">chrm_len</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int64</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">r_data</span> <span class="ow">in</span> <span class="n">cs_reads</span><span class="p">:</span>
        <span class="c1"># extract read means data so data across all chrms is not</span>
        <span class="c1"># in RAM at one time</span>
        <span class="n">g_slot_values</span> <span class="o">=</span> <span class="n">get_single_slot_genome_centric</span><span class="p">(</span><span class="n">r_data</span><span class="p">,</span> <span class="n">slot_name</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">g_slot_values</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span> <span class="k">continue</span>

        <span class="n">base_sums</span><span class="p">[</span><span class="n">r_data</span><span class="o">.</span><span class="n">start</span><span class="p">:</span>
                  <span class="n">r_data</span><span class="o">.</span><span class="n">start</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">g_slot_values</span><span class="p">)]</span> <span class="o">+=</span> <span class="n">g_slot_values</span>
        <span class="n">base_cov</span><span class="p">[</span><span class="n">r_data</span><span class="o">.</span><span class="n">start</span><span class="p">:</span><span class="n">r_data</span><span class="o">.</span><span class="n">start</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">g_slot_values</span><span class="p">)]</span> <span class="o">+=</span> <span class="mi">1</span>
</div>
    <span class="k">return</span> <span class="n">base_sums</span> <span class="o">/</span> <span class="n">base_cov</span>

<div class="viewcode-block" id="get_all_mean_slot_values"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.get_all_mean_slot_values">[docs]</a><span class="k">def</span> <span class="nf">get_all_mean_slot_values</span><span class="p">(</span><span class="n">raw_read_coverage</span><span class="p">,</span> <span class="n">chrm_sizes</span><span class="p">,</span> <span class="n">slot_name</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get the mean over all reads at each covered genomic location for this</span>
<span class="sd">    slots value over all covered chromosomes and strands</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># ignore divide by zero errors that occur where there is no</span>
    <span class="c1"># coverage. Need to correct nan values after subtracting two sets of</span>
    <span class="c1"># coverage so leave as nan for now</span>
    <span class="n">old_err_settings</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">seterr</span><span class="p">(</span><span class="nb">all</span><span class="o">=</span><span class="s1">&#39;ignore&#39;</span><span class="p">)</span>
    <span class="c1"># take the mean over all signal overlapping each base</span>
    <span class="n">all_mean_values</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span> <span class="ow">in</span> <span class="p">[(</span><span class="n">c</span><span class="p">,</span> <span class="n">s</span><span class="p">)</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">chrm_sizes</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="p">(</span><span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="s1">&#39;-&#39;</span><span class="p">)]:</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">)</span> <span class="ow">in</span> <span class="n">raw_read_coverage</span><span class="p">:</span>
            <span class="n">cs_mean_values</span> <span class="o">=</span> <span class="n">get_mean_slot_genome_centric</span><span class="p">(</span>
                <span class="n">raw_read_coverage</span><span class="p">[(</span><span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">)],</span> <span class="n">chrm_sizes</span><span class="p">[</span><span class="n">chrm</span><span class="p">],</span> <span class="n">slot_name</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">cs_mean_values</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">empty</span><span class="p">(</span><span class="n">chrm_sizes</span><span class="p">[</span><span class="n">chrm</span><span class="p">])</span>
            <span class="n">cs_mean_values</span><span class="p">[:]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>
        <span class="n">all_mean_values</span><span class="p">[(</span><span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">)]</span> <span class="o">=</span> <span class="n">cs_mean_values</span>
    <span class="n">_</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">seterr</span><span class="p">(</span><span class="o">**</span><span class="n">old_err_settings</span><span class="p">)</span>
</div>
    <span class="k">return</span> <span class="n">all_mean_values</span>

<div class="viewcode-block" id="get_all_mean_levels"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.get_all_mean_levels">[docs]</a><span class="k">def</span> <span class="nf">get_all_mean_levels</span><span class="p">(</span><span class="n">raw_read_coverage</span><span class="p">,</span> <span class="n">chrm_sizes</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Helper function to compute genome location mean levels</span>
<span class="sd">    &quot;&quot;&quot;</span></div>
    <span class="k">return</span> <span class="n">get_all_mean_slot_values</span><span class="p">(</span><span class="n">raw_read_coverage</span><span class="p">,</span> <span class="n">chrm_sizes</span><span class="p">,</span> <span class="s1">&#39;norm_mean&#39;</span><span class="p">)</span>

<div class="viewcode-block" id="get_all_mean_stdev"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.get_all_mean_stdev">[docs]</a><span class="k">def</span> <span class="nf">get_all_mean_stdev</span><span class="p">(</span><span class="n">raw_read_coverage</span><span class="p">,</span> <span class="n">chrm_sizes</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Helper function to compute genome location mean levels</span>
<span class="sd">    &quot;&quot;&quot;</span></div>
    <span class="k">return</span> <span class="n">get_all_mean_slot_values</span><span class="p">(</span><span class="n">raw_read_coverage</span><span class="p">,</span> <span class="n">chrm_sizes</span><span class="p">,</span> <span class="s1">&#39;norm_stdev&#39;</span><span class="p">)</span>

<div class="viewcode-block" id="get_all_mean_lengths"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.get_all_mean_lengths">[docs]</a><span class="k">def</span> <span class="nf">get_all_mean_lengths</span><span class="p">(</span><span class="n">raw_read_coverage</span><span class="p">,</span> <span class="n">chrm_sizes</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Helper function to compute genome location mean lengths</span>
<span class="sd">    &quot;&quot;&quot;</span></div>
    <span class="k">return</span> <span class="n">get_all_mean_slot_values</span><span class="p">(</span><span class="n">raw_read_coverage</span><span class="p">,</span> <span class="n">chrm_sizes</span><span class="p">,</span> <span class="s1">&#39;length&#39;</span><span class="p">)</span>


<span class="c1">###########################################</span>
<span class="c1">###### Special Data Access Functions ######</span>
<span class="c1">###########################################</span>

<div class="viewcode-block" id="get_channel_info"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.get_channel_info">[docs]</a><span class="k">def</span> <span class="nf">get_channel_info</span><span class="p">(</span><span class="n">read_data</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get channel information for a read</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">fast5_info</span> <span class="o">=</span> <span class="n">read_data</span><span class="p">[</span><span class="s1">&#39;UniqueGlobalKey/channel_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">attrs</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;No channel_id group in HDF5 file. &quot;</span> <span class="o">+</span>
                                  <span class="s2">&quot;Probably mux scan HDF5 file.&quot;</span><span class="p">)</span>

    <span class="n">channel_info</span> <span class="o">=</span> <span class="n">channelInfo</span><span class="p">(</span>
        <span class="n">fast5_info</span><span class="p">[</span><span class="s1">&#39;offset&#39;</span><span class="p">],</span> <span class="n">fast5_info</span><span class="p">[</span><span class="s1">&#39;range&#39;</span><span class="p">],</span>
        <span class="n">fast5_info</span><span class="p">[</span><span class="s1">&#39;digitisation&#39;</span><span class="p">],</span> <span class="n">fast5_info</span><span class="p">[</span><span class="s1">&#39;channel_number&#39;</span><span class="p">],</span>
        <span class="n">fast5_info</span><span class="p">[</span><span class="s1">&#39;sampling_rate&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">int64</span><span class="p">))</span>
</div>
    <span class="k">return</span> <span class="n">channel_info</span>

<div class="viewcode-block" id="get_raw_signal"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.get_raw_signal">[docs]</a><span class="k">def</span> <span class="nf">get_raw_signal</span><span class="p">(</span><span class="n">r_data</span><span class="p">,</span> <span class="n">int_start</span><span class="p">,</span> <span class="n">int_end</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Extract raw signal from where this read overlaps a particular genomic region</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">with</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="n">r_data</span><span class="o">.</span><span class="n">fn</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fast5_data</span><span class="p">:</span>
        <span class="c1"># retrieve shift and scale computed in correction script</span>
        <span class="n">corr_subgrp</span> <span class="o">=</span> <span class="n">fast5_data</span><span class="p">[</span><span class="s1">&#39;/Analyses/&#39;</span> <span class="o">+</span> <span class="n">r_data</span><span class="o">.</span><span class="n">corr_group</span><span class="p">]</span>
        <span class="n">event_starts</span> <span class="o">=</span> <span class="n">corr_subgrp</span><span class="p">[</span><span class="s1">&#39;Events&#39;</span><span class="p">][</span><span class="s1">&#39;start&#39;</span><span class="p">]</span>
        <span class="n">events_end</span> <span class="o">=</span> <span class="n">event_starts</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">corr_subgrp</span><span class="p">[</span><span class="s1">&#39;Events&#39;</span><span class="p">][</span><span class="s1">&#39;length&#39;</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">segs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">event_starts</span><span class="p">,</span> <span class="p">[</span><span class="n">events_end</span><span class="p">,]])</span>

        <span class="n">shift</span> <span class="o">=</span> <span class="n">corr_subgrp</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s1">&#39;shift&#39;</span><span class="p">]</span>
        <span class="n">scale</span> <span class="o">=</span> <span class="n">corr_subgrp</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s1">&#39;scale&#39;</span><span class="p">]</span>
        <span class="n">lower_lim</span> <span class="o">=</span> <span class="n">corr_subgrp</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s1">&#39;lower_lim&#39;</span><span class="p">]</span>
        <span class="n">upper_lim</span> <span class="o">=</span> <span class="n">corr_subgrp</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s1">&#39;upper_lim&#39;</span><span class="p">]</span>
        <span class="n">all_sig</span> <span class="o">=</span> <span class="n">get_raw_read_slot</span><span class="p">(</span><span class="n">fast5_data</span><span class="p">)[</span><span class="s1">&#39;Signal&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value</span>

    <span class="n">rsrtr</span> <span class="o">=</span> <span class="n">r_data</span><span class="o">.</span><span class="n">read_start_rel_to_raw</span>
    <span class="k">if</span> <span class="n">r_data</span><span class="o">.</span><span class="n">rna</span><span class="p">:</span>
        <span class="c1"># reverse raw signal for RNA</span>
        <span class="n">all_sig</span> <span class="o">=</span> <span class="n">all_sig</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

    <span class="k">if</span> <span class="n">r_data</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span><span class="p">:</span>
        <span class="n">segs</span> <span class="o">=</span> <span class="p">(</span><span class="n">segs</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span> <span class="o">+</span> <span class="n">segs</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

    <span class="k">if</span> <span class="n">int_start</span> <span class="o">&lt;</span> <span class="n">r_data</span><span class="o">.</span><span class="n">start</span><span class="p">:</span>
        <span class="c1"># handle reads that start in the middle of the interval</span>
        <span class="n">start_offset</span> <span class="o">=</span> <span class="n">r_data</span><span class="o">.</span><span class="n">start</span> <span class="o">-</span> <span class="n">int_start</span>
        <span class="n">overlap_seg_data</span> <span class="o">=</span> <span class="n">segs</span><span class="p">[:</span><span class="n">int_end</span> <span class="o">-</span> <span class="n">r_data</span><span class="o">.</span><span class="n">start</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">start_offset</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">skipped_bases</span> <span class="o">=</span> <span class="n">int_start</span> <span class="o">-</span> <span class="n">r_data</span><span class="o">.</span><span class="n">start</span>
        <span class="n">overlap_seg_data</span> <span class="o">=</span> <span class="n">segs</span><span class="p">[</span>
            <span class="n">skipped_bases</span><span class="p">:</span><span class="n">int_end</span> <span class="o">-</span> <span class="n">r_data</span><span class="o">.</span><span class="n">start</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]</span>

    <span class="n">num_reg_obs</span> <span class="o">=</span> <span class="n">overlap_seg_data</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">-</span> <span class="n">overlap_seg_data</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">r_data</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="s2">&quot;+&quot;</span><span class="p">:</span>
        <span class="n">reg_start_rel_raw</span> <span class="o">=</span> <span class="n">rsrtr</span> <span class="o">+</span> <span class="n">overlap_seg_data</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">r_sig</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">normalize_raw_signal</span><span class="p">(</span>
            <span class="n">all_sig</span><span class="p">,</span> <span class="n">reg_start_rel_raw</span><span class="p">,</span> <span class="n">num_reg_obs</span><span class="p">,</span> <span class="n">shift</span><span class="o">=</span><span class="n">shift</span><span class="p">,</span>
            <span class="n">scale</span><span class="o">=</span><span class="n">scale</span><span class="p">,</span> <span class="n">lower_lim</span><span class="o">=</span><span class="n">lower_lim</span><span class="p">,</span> <span class="n">upper_lim</span><span class="o">=</span><span class="n">upper_lim</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">reg_start_rel_raw</span> <span class="o">=</span> <span class="n">rsrtr</span> <span class="o">+</span> <span class="n">segs</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">-</span> <span class="n">overlap_seg_data</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">r_sig</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">normalize_raw_signal</span><span class="p">(</span>
            <span class="n">all_sig</span><span class="p">,</span> <span class="n">reg_start_rel_raw</span><span class="p">,</span> <span class="n">num_reg_obs</span><span class="p">,</span> <span class="n">shift</span><span class="o">=</span><span class="n">shift</span><span class="p">,</span>
            <span class="n">scale</span><span class="o">=</span><span class="n">scale</span><span class="p">,</span> <span class="n">lower_lim</span><span class="o">=</span><span class="n">lower_lim</span><span class="p">,</span> <span class="n">upper_lim</span><span class="o">=</span><span class="n">upper_lim</span><span class="p">)</span>
        <span class="n">r_sig</span> <span class="o">=</span> <span class="n">r_sig</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
</div>
    <span class="k">return</span> <span class="n">r_sig</span><span class="p">,</span> <span class="n">overlap_seg_data</span><span class="p">,</span> <span class="n">start_offset</span>

<div class="viewcode-block" id="parse_read_correction_data"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.parse_read_correction_data">[docs]</a><span class="k">def</span> <span class="nf">parse_read_correction_data</span><span class="p">(</span><span class="n">r_data</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Parse correction data from an event resquiggled read</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="k">with</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="n">r_data</span><span class="o">.</span><span class="n">fn</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fast5_data</span><span class="p">:</span>
            <span class="n">corr_grp</span> <span class="o">=</span> <span class="n">fast5_data</span><span class="p">[</span><span class="s1">&#39;/Analyses/&#39;</span> <span class="o">+</span> <span class="n">r_data</span><span class="o">.</span><span class="n">corr_group</span><span class="p">]</span>
            <span class="n">events_grp</span> <span class="o">=</span> <span class="n">corr_grp</span><span class="p">[</span><span class="s1">&#39;Events&#39;</span><span class="p">]</span>
            <span class="n">event_starts</span> <span class="o">=</span> <span class="n">events_grp</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span>
            <span class="n">events_end</span> <span class="o">=</span> <span class="n">event_starts</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">events_grp</span><span class="p">[</span><span class="s1">&#39;length&#39;</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
            <span class="n">new_segs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">event_starts</span><span class="p">,</span> <span class="p">[</span><span class="n">events_end</span><span class="p">,]])</span>

            <span class="n">raw_grp</span> <span class="o">=</span> <span class="n">get_raw_read_slot</span><span class="p">(</span><span class="n">fast5_data</span><span class="p">)</span>
            <span class="n">read_id</span> <span class="o">=</span> <span class="n">raw_grp</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s1">&#39;read_id&#39;</span><span class="p">]</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">read_id</span> <span class="o">=</span> <span class="n">read_id</span><span class="o">.</span><span class="n">decode</span><span class="p">()</span>
            <span class="k">except</span> <span class="p">(</span><span class="ne">AttributeError</span><span class="p">,</span> <span class="ne">TypeError</span><span class="p">):</span>
                <span class="k">pass</span>
            <span class="n">signal_data</span> <span class="o">=</span> <span class="n">raw_grp</span><span class="p">[</span><span class="s1">&#39;Signal&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value</span>

            <span class="n">raw_offset</span> <span class="o">=</span> <span class="n">events_grp</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s1">&#39;read_start_rel_to_raw&#39;</span><span class="p">]</span>
            <span class="n">shift</span><span class="p">,</span> <span class="n">scale</span><span class="p">,</span> <span class="n">lower_lim</span><span class="p">,</span> <span class="n">upper_lim</span> <span class="o">=</span> <span class="p">[</span>
                <span class="n">corr_grp</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="n">attr_name</span><span class="p">]</span> <span class="k">for</span> <span class="n">attr_name</span> <span class="ow">in</span> <span class="p">(</span>
                    <span class="s1">&#39;shift&#39;</span><span class="p">,</span> <span class="s1">&#39;scale&#39;</span><span class="p">,</span> <span class="s1">&#39;lower_lim&#39;</span><span class="p">,</span> <span class="s1">&#39;upper_lim&#39;</span><span class="p">)]</span>

            <span class="n">old_segs</span> <span class="o">=</span> <span class="n">corr_grp</span><span class="p">[</span><span class="s1">&#39;Alignment/read_segments&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value</span>
            <span class="n">old_align_vals</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span>
                <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">decode</span><span class="p">(),</span>
                <span class="n">corr_grp</span><span class="p">[</span><span class="s1">&#39;Alignment/read_alignment&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value</span><span class="p">))</span>
            <span class="n">new_align_vals</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span>
                <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">decode</span><span class="p">(),</span>
                <span class="n">corr_grp</span><span class="p">[</span><span class="s1">&#39;Alignment/genome_alignment&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value</span><span class="p">))</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="k">if</span> <span class="n">r_data</span><span class="o">.</span><span class="n">rna</span><span class="p">:</span>
        <span class="n">signal_data</span> <span class="o">=</span> <span class="n">signal_data</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

    <span class="k">return</span> <span class="p">(</span><span class="n">read_id</span><span class="p">,</span> <span class="n">signal_data</span><span class="p">,</span> <span class="n">raw_offset</span><span class="p">,</span> <span class="n">shift</span><span class="p">,</span> <span class="n">scale</span><span class="p">,</span> <span class="n">lower_lim</span><span class="p">,</span>
            <span class="n">upper_lim</span><span class="p">,</span> <span class="n">old_segs</span><span class="p">,</span> <span class="n">old_align_vals</span><span class="p">,</span> <span class="n">new_align_vals</span><span class="p">,</span></div>
            <span class="n">events_end</span><span class="p">,</span> <span class="n">new_segs</span><span class="p">)</span>

<div class="viewcode-block" id="get_all_read_data"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.get_all_read_data">[docs]</a><span class="k">def</span> <span class="nf">get_all_read_data</span><span class="p">(</span><span class="n">r_data</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Extract most relevant read data from this read</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="k">with</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="n">r_data</span><span class="o">.</span><span class="n">fn</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fast5_data</span><span class="p">:</span>
            <span class="c1"># note that it&#39;s more efficient to try to access the slot</span>
            <span class="c1"># and except the error that check if the slot exists first</span>
            <span class="n">corr_subgrp</span> <span class="o">=</span> <span class="n">fast5_data</span><span class="p">[</span><span class="s1">&#39;/Analyses/&#39;</span> <span class="o">+</span> <span class="n">r_data</span><span class="o">.</span><span class="n">corr_group</span><span class="p">]</span>
            <span class="n">algn_subgrp</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">corr_subgrp</span><span class="p">[</span><span class="s1">&#39;Alignment&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">attrs</span><span class="o">.</span><span class="n">items</span><span class="p">())</span>
            <span class="n">event_data</span> <span class="o">=</span> <span class="n">corr_subgrp</span><span class="p">[</span><span class="s1">&#39;Events&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value</span>
            <span class="n">r_attrs</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">corr_subgrp</span><span class="o">.</span><span class="n">attrs</span><span class="o">.</span><span class="n">items</span><span class="p">())</span>
            <span class="n">all_sig</span> <span class="o">=</span> <span class="n">get_raw_read_slot</span><span class="p">(</span><span class="n">fast5_data</span><span class="p">)[</span><span class="s1">&#39;Signal&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="c1"># probably truncated file or Events slot doesn&#39;t exist</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="k">if</span> <span class="n">r_data</span><span class="o">.</span><span class="n">rna</span><span class="p">:</span>
        <span class="n">all_sig</span> <span class="o">=</span> <span class="n">all_sig</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">r_means</span> <span class="o">=</span> <span class="n">event_data</span><span class="p">[</span><span class="s1">&#39;norm_mean&#39;</span><span class="p">]</span>
    <span class="n">r_seq</span> <span class="o">=</span> <span class="sa">b</span><span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">event_data</span><span class="p">[</span><span class="s1">&#39;base&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">decode</span><span class="p">()</span>

    <span class="n">events_end</span> <span class="o">=</span> <span class="n">event_data</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">event_data</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;length&#39;</span><span class="p">]</span>
    <span class="n">segs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">event_data</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">],</span> <span class="p">[</span><span class="n">events_end</span><span class="p">,]])</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">int64</span><span class="p">)</span>
    <span class="n">r_sig</span><span class="p">,</span> <span class="n">scale_values</span> <span class="o">=</span> <span class="n">normalize_raw_signal</span><span class="p">(</span>
        <span class="n">all_sig</span><span class="p">,</span> <span class="n">r_data</span><span class="o">.</span><span class="n">read_start_rel_to_raw</span><span class="p">,</span> <span class="n">segs</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">-</span> <span class="n">segs</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
        <span class="n">shift</span><span class="o">=</span><span class="n">r_attrs</span><span class="p">[</span><span class="s1">&#39;shift&#39;</span><span class="p">],</span> <span class="n">scale</span><span class="o">=</span><span class="n">r_attrs</span><span class="p">[</span><span class="s1">&#39;scale&#39;</span><span class="p">],</span>
        <span class="n">lower_lim</span><span class="o">=</span><span class="n">r_attrs</span><span class="p">[</span><span class="s1">&#39;lower_lim&#39;</span><span class="p">],</span> <span class="n">upper_lim</span><span class="o">=</span><span class="n">r_attrs</span><span class="p">[</span><span class="s1">&#39;upper_lim&#39;</span><span class="p">])</span>

    <span class="k">return</span> <span class="p">(</span><span class="n">r_means</span><span class="p">,</span> <span class="n">r_seq</span><span class="p">,</span> <span class="n">r_sig</span><span class="p">,</span> <span class="n">segs</span><span class="p">,</span> <span class="n">scale_values</span><span class="p">,</span>
            <span class="n">r_attrs</span><span class="p">[</span><span class="s1">&#39;norm_type&#39;</span><span class="p">],</span> <span class="n">r_attrs</span><span class="p">[</span><span class="s1">&#39;outlier_threshold&#39;</span><span class="p">],</span>
            <span class="n">genomeLoc</span><span class="p">(</span><span class="n">algn_subgrp</span><span class="p">[</span><span class="s1">&#39;mapped_start&#39;</span><span class="p">],</span> <span class="n">algn_subgrp</span><span class="p">[</span><span class="s1">&#39;mapped_strand&#39;</span><span class="p">],</span></div>
                      <span class="n">algn_subgrp</span><span class="p">[</span><span class="s1">&#39;mapped_chrom&#39;</span><span class="p">]))</span>

<div class="viewcode-block" id="get_coverage"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.get_coverage">[docs]</a><span class="k">def</span> <span class="nf">get_coverage</span><span class="p">(</span><span class="n">raw_read_coverage</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get genome coverage for a set of reads</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">VERBOSE</span><span class="p">:</span> <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;Calculating read coverage.</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="n">read_coverage</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">),</span> <span class="n">reads_data</span> <span class="ow">in</span> <span class="n">raw_read_coverage</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">reads_data</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span> <span class="k">continue</span>
        <span class="n">max_end</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">r_data</span><span class="o">.</span><span class="n">end</span> <span class="k">for</span> <span class="n">r_data</span> <span class="ow">in</span> <span class="n">reads_data</span><span class="p">)</span>
        <span class="n">chrm_coverage</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">max_end</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int64</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">r_data</span> <span class="ow">in</span> <span class="n">reads_data</span><span class="p">:</span>
            <span class="n">chrm_coverage</span><span class="p">[</span><span class="n">r_data</span><span class="o">.</span><span class="n">start</span><span class="p">:</span><span class="n">r_data</span><span class="o">.</span><span class="n">end</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="n">read_coverage</span><span class="p">[(</span><span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">)]</span> <span class="o">=</span> <span class="n">chrm_coverage</span>
</div>
    <span class="k">return</span> <span class="n">read_coverage</span>

<div class="viewcode-block" id="get_reads_events"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.get_reads_events">[docs]</a><span class="k">def</span> <span class="nf">get_reads_events</span><span class="p">(</span><span class="n">cs_reads</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Extract read base levels split by genomic position</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># note that this function assumes that all reads come from the same</span>
    <span class="c1"># chromosome and strand</span>
    <span class="n">cs_base_means</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">cs_read_start_ends</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">r_data</span> <span class="ow">in</span> <span class="n">cs_reads</span><span class="p">:</span>
        <span class="c1"># extract read means data so data across all chrms is not</span>
        <span class="c1"># in RAM at one time</span>
        <span class="n">read_means</span> <span class="o">=</span> <span class="n">get_single_slot_genome_centric</span><span class="p">(</span><span class="n">r_data</span><span class="p">,</span> <span class="s1">&#39;norm_mean&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">read_means</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span> <span class="k">continue</span>
        <span class="k">assert</span> <span class="n">read_means</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="n">r_data</span><span class="o">.</span><span class="n">end</span> <span class="o">-</span> <span class="n">r_data</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="p">(</span>
            <span class="s1">&#39;Read found with mismatching mapping location and &#39;</span> <span class="o">+</span>
            <span class="s1">&#39;signal information.&#39;</span><span class="p">)</span>
        <span class="n">cs_base_means</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">read_means</span><span class="p">)</span>
        <span class="n">cs_read_start_ends</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">r_data</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">r_data</span><span class="o">.</span><span class="n">end</span><span class="p">))</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cs_base_means</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span> <span class="k">return</span> <span class="kc">None</span>

    <span class="n">chrm_signal</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span><span class="n">cs_base_means</span><span class="p">)</span>
    <span class="n">chrm_pos</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">)</span>
                               <span class="k">for</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="ow">in</span> <span class="n">cs_read_start_ends</span><span class="p">])</span>
    <span class="c1"># get order of all bases from position array</span>
    <span class="n">as_chrm_pos</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">chrm_pos</span><span class="p">)</span>
    <span class="c1"># then sort the signal array by genomic position and</span>
    <span class="c1"># split into event means by base</span>
    <span class="n">split_poss</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span>
        <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([[</span><span class="mi">0</span><span class="p">,],</span> <span class="n">np</span><span class="o">.</span><span class="n">diff</span><span class="p">(</span>
            <span class="n">chrm_pos</span><span class="p">[</span><span class="n">as_chrm_pos</span><span class="p">])])</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">cs_base_events</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span>
        <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">chrm_pos</span><span class="p">[</span><span class="n">as_chrm_pos</span><span class="p">]),</span>
        <span class="n">np</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">chrm_signal</span><span class="p">[</span><span class="n">as_chrm_pos</span><span class="p">],</span> <span class="n">split_poss</span><span class="p">)))</span>
</div>
    <span class="k">return</span> <span class="n">cs_base_events</span>

<div class="viewcode-block" id="update_seq"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.update_seq">[docs]</a><span class="k">def</span> <span class="nf">update_seq</span><span class="p">(</span><span class="n">r_data</span><span class="p">,</span> <span class="n">reg_base_data</span><span class="p">,</span> <span class="n">int_start</span><span class="p">,</span> <span class="n">int_end</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Update the sequence for the region based on this read</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">r_seq</span> <span class="o">=</span> <span class="sa">b</span><span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">get_single_slot_read_centric</span><span class="p">(</span><span class="n">r_data</span><span class="p">,</span> <span class="s1">&#39;base&#39;</span><span class="p">))</span><span class="o">.</span><span class="n">decode</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">r_seq</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="c1"># probably a corrupt file so return that the region is only</span>
        <span class="c1"># up to the start of this read so the next valid read will be added</span>
        <span class="c1"># to the sequence</span>
        <span class="k">return</span> <span class="n">reg_base_data</span><span class="p">,</span> <span class="nb">max</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">r_data</span><span class="o">.</span><span class="n">start</span> <span class="o">-</span> <span class="n">int_start</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">r_data</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;-&#39;</span><span class="p">:</span>
        <span class="n">r_seq</span> <span class="o">=</span> <span class="n">rev_comp</span><span class="p">(</span><span class="n">r_seq</span><span class="p">)</span>

    <span class="c1"># if read starts before the interval</span>
    <span class="k">if</span> <span class="n">r_data</span><span class="o">.</span><span class="n">start</span> <span class="o">&lt;=</span> <span class="n">int_start</span><span class="p">:</span>
        <span class="n">r_end_overlap</span> <span class="o">=</span> <span class="n">r_data</span><span class="o">.</span><span class="n">end</span> <span class="o">-</span> <span class="n">int_start</span>
        <span class="c1"># if read covers the whole interval</span>
        <span class="k">if</span> <span class="n">r_data</span><span class="o">.</span><span class="n">end</span> <span class="o">&gt;</span> <span class="n">int_end</span><span class="p">:</span>
            <span class="n">r_end_clip</span> <span class="o">=</span> <span class="n">r_data</span><span class="o">.</span><span class="n">end</span> <span class="o">-</span> <span class="n">int_end</span>
            <span class="n">reg_base_data</span> <span class="o">=</span> <span class="n">r_seq</span><span class="p">[</span><span class="o">-</span><span class="n">r_end_overlap</span><span class="p">:</span><span class="o">-</span><span class="n">r_end_clip</span><span class="p">]</span>
            <span class="k">return</span> <span class="n">reg_base_data</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">reg_base_data</span><span class="p">)</span>
        <span class="c1"># end of read overlaps beginning of interval</span>
        <span class="n">reg_base_data</span><span class="p">[:</span><span class="n">r_end_overlap</span><span class="p">]</span> <span class="o">=</span> <span class="n">r_seq</span><span class="p">[</span><span class="o">-</span><span class="n">r_end_overlap</span><span class="p">:]</span>
        <span class="k">return</span> <span class="n">reg_base_data</span><span class="p">,</span> <span class="n">r_end_overlap</span>
    <span class="c1"># read doesn&#39;t cover the beginning of region</span>
    <span class="k">if</span> <span class="n">r_data</span><span class="o">.</span><span class="n">end</span> <span class="o">&gt;</span> <span class="n">int_end</span><span class="p">:</span>
        <span class="c1"># beginning of read covers to the end of the region</span>
        <span class="n">r_begin_overlap</span> <span class="o">=</span> <span class="n">int_end</span> <span class="o">-</span> <span class="n">r_data</span><span class="o">.</span><span class="n">start</span>
        <span class="n">reg_base_data</span><span class="p">[</span><span class="o">-</span><span class="n">r_begin_overlap</span><span class="p">:]</span> <span class="o">=</span> <span class="n">r_seq</span><span class="p">[:</span><span class="n">r_begin_overlap</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">reg_base_data</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">reg_base_data</span><span class="p">)</span>
    <span class="c1"># first read is completely contained in the interval</span>
    <span class="n">r_len</span> <span class="o">=</span> <span class="n">r_data</span><span class="o">.</span><span class="n">end</span> <span class="o">-</span> <span class="n">r_data</span><span class="o">.</span><span class="n">start</span>
    <span class="n">r_int_start</span> <span class="o">=</span> <span class="n">r_data</span><span class="o">.</span><span class="n">start</span> <span class="o">-</span> <span class="n">int_start</span>
    <span class="n">reg_base_data</span><span class="p">[</span><span class="n">r_int_start</span><span class="p">:</span><span class="n">r_int_start</span> <span class="o">+</span> <span class="n">r_len</span><span class="p">]</span> <span class="o">=</span> <span class="n">r_seq</span></div>
    <span class="k">return</span> <span class="n">reg_base_data</span><span class="p">,</span> <span class="n">r_int_start</span> <span class="o">+</span> <span class="n">r_len</span>

<div class="viewcode-block" id="get_seq_from_reads"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.get_seq_from_reads">[docs]</a><span class="k">def</span> <span class="nf">get_seq_from_reads</span><span class="p">(</span><span class="n">int_start</span><span class="p">,</span> <span class="n">int_end</span><span class="p">,</span> <span class="n">reg_reads</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Extract the forward strand genomic sequence for an interval from</span>
<span class="sd">    a set of reads</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># handle case where no read overlaps whole region</span>
    <span class="c1"># let each read contibute its sequence and fill the rest</span>
    <span class="c1"># with dashes</span>
    <span class="n">reg_base_data</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;-&#39;</span><span class="p">]</span> <span class="o">*</span> <span class="p">(</span><span class="n">int_end</span> <span class="o">-</span> <span class="n">int_start</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">reg_reads</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">return</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">reg_base_data</span><span class="p">)</span>
    <span class="c1"># get region sequence by moving through reads that</span>
    <span class="c1"># cover the region, but only extract seqeunce from the</span>
    <span class="c1"># (close to) minimal reads</span>
    <span class="n">s_reg_reads</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">reg_reads</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">r</span><span class="p">:</span> <span class="p">(</span><span class="n">r</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">r</span><span class="o">.</span><span class="n">end</span><span class="p">))</span>
    <span class="c1"># begin filling sequence with first (by start pos) read</span>
    <span class="n">reg_base_data</span><span class="p">,</span> <span class="n">curr_cov_pos</span> <span class="o">=</span> <span class="n">update_seq</span><span class="p">(</span>
        <span class="n">s_reg_reads</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="mi">0</span><span class="p">),</span> <span class="n">reg_base_data</span><span class="p">,</span> <span class="n">int_start</span><span class="p">,</span> <span class="n">int_end</span><span class="p">)</span>
    <span class="c1"># if there was only one read return now</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">s_reg_reads</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">curr_cov_pos</span> <span class="o">&gt;=</span> <span class="n">int_end</span> <span class="o">-</span> <span class="n">int_start</span><span class="p">:</span>
        <span class="k">return</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">reg_base_data</span><span class="p">)</span>

    <span class="c1"># get next read (by start pos)</span>
    <span class="n">curr_read</span> <span class="o">=</span> <span class="n">s_reg_reads</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">next_read</span> <span class="ow">in</span> <span class="n">s_reg_reads</span><span class="p">:</span>
        <span class="c1"># once the next read start passes the region covered thus far</span>
        <span class="c1"># add the sequence for the saved curr_read to the reg sequence</span>
        <span class="k">if</span> <span class="n">next_read</span><span class="o">.</span><span class="n">start</span> <span class="o">&gt;=</span> <span class="n">curr_cov_pos</span><span class="p">:</span>
            <span class="c1"># add read with curr longest end position to the region seq</span>
            <span class="n">reg_base_data</span><span class="p">,</span> <span class="n">curr_cov_pos</span> <span class="o">=</span> <span class="n">update_seq</span><span class="p">(</span>
                <span class="n">curr_read</span><span class="p">,</span> <span class="n">reg_base_data</span><span class="p">,</span> <span class="n">int_start</span><span class="p">,</span> <span class="n">int_end</span><span class="p">)</span>
            <span class="n">curr_read</span> <span class="o">=</span> <span class="n">next_read</span>
            <span class="c1"># if the whole interval is covered return the sequence</span>
            <span class="k">if</span> <span class="n">curr_cov_pos</span> <span class="o">&gt;=</span> <span class="n">int_end</span> <span class="o">-</span> <span class="n">int_start</span><span class="p">:</span>
                <span class="k">return</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">reg_base_data</span><span class="p">)</span>
            <span class="k">continue</span>
        <span class="k">if</span> <span class="n">next_read</span><span class="o">.</span><span class="n">end</span> <span class="o">&gt;</span> <span class="n">curr_read</span><span class="o">.</span><span class="n">end</span><span class="p">:</span>
            <span class="n">curr_read</span> <span class="o">=</span> <span class="n">next_read</span>

    <span class="n">reg_base_data</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">update_seq</span><span class="p">(</span>
        <span class="n">curr_read</span><span class="p">,</span> <span class="n">reg_base_data</span><span class="p">,</span> <span class="n">int_start</span><span class="p">,</span> <span class="n">int_end</span><span class="p">)</span>
</div>
    <span class="k">return</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">reg_base_data</span><span class="p">)</span>

<div class="viewcode-block" id="add_reg_seq"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.add_reg_seq">[docs]</a><span class="k">def</span> <span class="nf">add_reg_seq</span><span class="p">(</span><span class="n">all_reg_data</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Add the region sequence to the region data by extraction from a minimal</span>
<span class="sd">    set of reads</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">all_reg_base_data</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">reg_data</span> <span class="ow">in</span> <span class="n">all_reg_data</span><span class="p">:</span>
        <span class="c1"># add text to each regions data</span>
        <span class="n">all_reg_base_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">reg_data</span><span class="o">.</span><span class="n">_replace</span><span class="p">(</span><span class="n">seq</span><span class="o">=</span><span class="n">get_seq_from_reads</span><span class="p">(</span>
            <span class="n">reg_data</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">reg_data</span><span class="o">.</span><span class="n">end</span><span class="p">,</span> <span class="n">reg_data</span><span class="o">.</span><span class="n">reads</span><span class="p">)))</span>
</div>
    <span class="k">return</span> <span class="n">all_reg_base_data</span>

<div class="viewcode-block" id="get_region_reads"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.get_region_reads">[docs]</a><span class="k">def</span> <span class="nf">get_region_reads</span><span class="p">(</span>
        <span class="n">plot_intervals</span><span class="p">,</span> <span class="n">raw_read_coverage</span><span class="p">,</span> <span class="n">filter_no_cov</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">add_seq</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get all reads overlapping a set of intervals</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">get_c_s_data</span><span class="p">(</span><span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">):</span>
        <span class="c1"># get all reads intersecting the interval</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">)</span> <span class="ow">in</span> <span class="n">raw_read_coverage</span><span class="p">:</span>
            <span class="k">return</span> <span class="p">[</span>
                <span class="n">r_data</span> <span class="k">for</span> <span class="n">r_data</span> <span class="ow">in</span> <span class="n">raw_read_coverage</span><span class="p">[(</span><span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">)]</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="n">r_data</span><span class="o">.</span><span class="n">start</span> <span class="o">&gt;=</span> <span class="n">end</span> <span class="ow">or</span> <span class="n">r_data</span><span class="o">.</span><span class="n">end</span> <span class="o">&lt;=</span> <span class="n">start</span><span class="p">)]</span>
        <span class="k">return</span> <span class="p">[]</span>


    <span class="n">all_reg_data</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">int_i</span> <span class="ow">in</span> <span class="n">plot_intervals</span><span class="p">:</span>
        <span class="c1"># get all reads that overlap this interval</span>
        <span class="c1"># note that this includes partial overlaps as these contribute</span>
        <span class="c1"># to coverage and other statistics so can&#39;t really restrict to</span>
        <span class="c1"># full coverage as previous versions of code did</span>
        <span class="k">if</span> <span class="n">int_i</span><span class="o">.</span><span class="n">strand</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="c1"># if strand is None, get data from both strands</span>
            <span class="n">all_reg_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">int_i</span><span class="o">.</span><span class="n">_replace</span><span class="p">(</span>
                <span class="n">reads</span><span class="o">=</span><span class="n">get_c_s_data</span><span class="p">(</span><span class="n">int_i</span><span class="o">.</span><span class="n">chrm</span><span class="p">,</span> <span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="n">int_i</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">int_i</span><span class="o">.</span><span class="n">end</span><span class="p">)</span> <span class="o">+</span>
                <span class="n">get_c_s_data</span><span class="p">(</span><span class="n">int_i</span><span class="o">.</span><span class="n">chrm</span><span class="p">,</span> <span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">int_i</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">int_i</span><span class="o">.</span><span class="n">end</span><span class="p">)))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">all_reg_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">int_i</span><span class="o">.</span><span class="n">_replace</span><span class="p">(</span>
                <span class="n">reads</span><span class="o">=</span><span class="n">get_c_s_data</span><span class="p">(</span><span class="n">int_i</span><span class="o">.</span><span class="n">chrm</span><span class="p">,</span> <span class="n">int_i</span><span class="o">.</span><span class="n">strand</span><span class="p">,</span>
                                   <span class="n">int_i</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">int_i</span><span class="o">.</span><span class="n">end</span><span class="p">)))</span>

    <span class="k">if</span> <span class="n">add_seq</span><span class="p">:</span>
        <span class="n">all_reg_data</span> <span class="o">=</span> <span class="n">add_reg_seq</span><span class="p">(</span><span class="n">all_reg_data</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">filter_no_cov</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">all_reg_data</span>

    <span class="c1"># filter out no coverage regions</span>
    <span class="n">all_reg_data</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">reg_data</span> <span class="k">for</span> <span class="n">reg_data</span> <span class="ow">in</span> <span class="n">all_reg_data</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">reg_data</span><span class="o">.</span><span class="n">reads</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">]</span>

    <span class="n">no_cov_regions</span> <span class="o">=</span> <span class="p">[</span>
        <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">reg_data</span><span class="o">.</span><span class="n">reads</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">,</span> <span class="n">unicode</span><span class="p">(</span><span class="n">reg_data</span><span class="o">.</span><span class="n">chrm</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;:&#39;</span> <span class="o">+</span>
         <span class="n">unicode</span><span class="p">(</span><span class="n">reg_data</span><span class="o">.</span><span class="n">start</span><span class="p">))</span>
        <span class="k">for</span> <span class="n">reg_data</span> <span class="ow">in</span> <span class="n">all_reg_data</span><span class="p">]</span>
    <span class="k">if</span> <span class="nb">any</span><span class="p">(</span><span class="n">no_cov</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">no_cov</span> <span class="ow">in</span> <span class="n">no_cov_regions</span><span class="p">):</span>
        <span class="n">_warning_message</span><span class="p">(</span>
            <span class="s1">&#39;No coverage in regions: &#39;</span> <span class="o">+</span> <span class="s1">&#39;; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span>
                <span class="n">reg</span> <span class="k">for</span> <span class="n">no_cov</span><span class="p">,</span> <span class="n">reg</span> <span class="ow">in</span> <span class="n">no_cov_regions</span> <span class="k">if</span> <span class="n">no_cov</span><span class="p">]))</span>
</div>
    <span class="k">return</span> <span class="n">all_reg_data</span>

<div class="viewcode-block" id="get_region_sequences"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.get_region_sequences">[docs]</a><span class="k">def</span> <span class="nf">get_region_sequences</span><span class="p">(</span>
        <span class="n">plot_intervals</span><span class="p">,</span> <span class="n">raw_read_coverage1</span><span class="p">,</span> <span class="n">raw_read_coverage2</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get the sequence for a set of intervals from a set of reads</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">all_reg_data</span> <span class="o">=</span> <span class="n">get_region_reads</span><span class="p">(</span>
        <span class="n">plot_intervals</span><span class="p">,</span> <span class="n">raw_read_coverage1</span><span class="p">,</span> <span class="n">filter_no_cov</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">add_seq</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">raw_read_coverage2</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">all_reg_data2</span> <span class="o">=</span> <span class="n">get_region_reads</span><span class="p">(</span>
            <span class="n">plot_intervals</span><span class="p">,</span> <span class="n">raw_read_coverage2</span><span class="p">,</span> <span class="n">filter_no_cov</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="n">add_seq</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">all_reg_data</span> <span class="o">=</span> <span class="p">[</span><span class="n">r1</span><span class="o">.</span><span class="n">_replace</span><span class="p">(</span><span class="n">reads</span><span class="o">=</span><span class="n">r1</span><span class="o">.</span><span class="n">reads</span> <span class="o">+</span> <span class="n">r2</span><span class="o">.</span><span class="n">reads</span><span class="p">)</span>
                        <span class="k">for</span> <span class="n">r1</span><span class="p">,</span> <span class="n">r2</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">all_reg_data</span><span class="p">,</span> <span class="n">all_reg_data2</span><span class="p">)]</span>
    <span class="n">all_reg_data</span> <span class="o">=</span> <span class="n">add_reg_seq</span><span class="p">(</span><span class="n">all_reg_data</span><span class="p">)</span>
</div>
    <span class="k">return</span> <span class="n">all_reg_data</span>


<span class="c1">###################################</span>
<span class="c1">###### FAST5 Write Functions ######</span>
<span class="c1">###################################</span>

<div class="viewcode-block" id="prep_fast5"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.prep_fast5">[docs]</a><span class="k">def</span> <span class="nf">prep_fast5</span><span class="p">(</span><span class="n">fast5_fn</span><span class="p">,</span> <span class="n">corr_grp</span><span class="p">,</span> <span class="n">overwrite</span><span class="p">,</span> <span class="n">in_place</span><span class="p">,</span>
               <span class="n">bc_grp</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">return_fp</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Prepare a read for re-squiggle processing (This deletes old re-squiggle</span>
<span class="sd">    info for this read)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">try_close_prep_err</span><span class="p">(</span><span class="n">fast5_data</span><span class="p">,</span> <span class="n">err_str</span><span class="p">):</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">fast5_data</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="k">pass</span>
        <span class="k">return</span> <span class="p">(</span><span class="n">err_str</span><span class="p">,</span> <span class="n">fast5_fn</span><span class="p">)</span>

    <span class="c1"># several checks to prepare the FAST5 file for correction before</span>
    <span class="c1"># processing to save compute</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">in_place</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">(</span><span class="s1">&#39;Not currently implementing new hdf5 file writing&#39;</span><span class="p">,</span>
                <span class="n">fast5_fn</span><span class="p">)</span>
    <span class="c1"># check that the file is writeable before trying to correct</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">access</span><span class="p">(</span><span class="n">fast5_fn</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">W_OK</span><span class="p">):</span>
        <span class="k">return</span> <span class="p">(</span><span class="s1">&#39;FAST5 file is not writable&#39;</span><span class="p">,</span> <span class="n">fast5_fn</span><span class="p">)</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="c1"># create group to store data</span>
        <span class="n">fast5_data</span> <span class="o">=</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="n">fast5_fn</span><span class="p">,</span> <span class="s1">&#39;r+&#39;</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">analyses_grp</span> <span class="o">=</span> <span class="n">fast5_data</span><span class="p">[</span><span class="s1">&#39;/Analyses&#39;</span><span class="p">]</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">try_close_prep_err</span><span class="p">(</span>
                <span class="n">fast5_data</span><span class="p">,</span> <span class="s1">&#39;Analysis group not found at root of FAST5&#39;</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="c1"># check that the requested basecalls group exsists</span>
            <span class="k">if</span> <span class="n">bc_grp</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">analyses_grp</span><span class="p">[</span><span class="n">bc_grp</span><span class="p">]</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">try_close_prep_err</span><span class="p">(</span>
                <span class="n">fast5_data</span><span class="p">,</span> <span class="s1">&#39;Basecall group not found at [--basecall-group]&#39;</span><span class="p">)</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">corr_grp_ptr</span> <span class="o">=</span> <span class="n">analyses_grp</span><span class="p">[</span><span class="n">corr_grp</span><span class="p">]</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">overwrite</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">try_close_prep_err</span><span class="p">(</span>
                    <span class="n">fast5_data</span><span class="p">,</span> <span class="s2">&quot;Tombo data exsists in [--corrected-group] &quot;</span> <span class="o">+</span>
                    <span class="s2">&quot;and [--overwrite] is not set&quot;</span><span class="p">)</span>
            <span class="k">del</span> <span class="n">analyses_grp</span><span class="p">[</span><span class="n">corr_grp</span><span class="p">]</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="c1"># if the corr_grp isn&#39;t there we will write it now, but</span>
            <span class="c1"># it&#39;s more efficient to try than to check if the slot is there</span>
            <span class="k">pass</span>

        <span class="n">corr_grp</span> <span class="o">=</span> <span class="n">analyses_grp</span><span class="o">.</span><span class="n">create_group</span><span class="p">(</span><span class="n">corr_grp</span><span class="p">)</span>
        <span class="n">corr_grp</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s1">&#39;tombo_version&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">TOMBO_VERSION</span>
        <span class="n">corr_grp</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s1">&#39;basecall_group&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">bc_grp</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">(</span>
            <span class="s1">&#39;Error opening or writing to fast5 file&#39;</span><span class="p">,</span> <span class="n">fast5_fn</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">return_fp</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">fast5_data</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="n">fast5_data</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="k">return</span> <span class="s1">&#39;Error closing fast5 file&#39;</span><span class="p">,</span> <span class="n">fast5_fn</span>
</div>
    <span class="k">return</span>

<div class="viewcode-block" id="write_error_status"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.write_error_status">[docs]</a><span class="k">def</span> <span class="nf">write_error_status</span><span class="p">(</span>
        <span class="n">filename</span><span class="p">,</span> <span class="n">corrected_group</span><span class="p">,</span> <span class="n">basecall_subgroup</span><span class="p">,</span> <span class="n">error_text</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Write error message for a read into the FAST5 file</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">fast5_data</span> <span class="o">=</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s1">&#39;r+&#39;</span><span class="p">)</span>
    <span class="n">analysis_grp</span> <span class="o">=</span> <span class="n">fast5_data</span><span class="p">[</span><span class="s1">&#39;/Analyses&#39;</span><span class="p">]</span>
    <span class="n">corr_grp</span> <span class="o">=</span> <span class="n">analysis_grp</span><span class="p">[</span><span class="n">corrected_group</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">basecall_subgroup</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="c1"># add subgroup matching subgroup from original basecalls</span>
        <span class="n">corr_subgrp</span> <span class="o">=</span> <span class="n">corr_grp</span><span class="o">.</span><span class="n">create_group</span><span class="p">(</span><span class="n">basecall_subgroup</span><span class="p">)</span>
        <span class="n">corr_subgrp</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s1">&#39;status&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">error_text</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">corr_grp</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s1">&#39;status&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">error_text</span>
</div>
    <span class="k">return</span>

<div class="viewcode-block" id="write_new_fast5_group"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.write_new_fast5_group">[docs]</a><span class="k">def</span> <span class="nf">write_new_fast5_group</span><span class="p">(</span>
        <span class="n">fast5_data</span><span class="p">,</span> <span class="n">genome_location</span><span class="p">,</span> <span class="n">read_start_rel_to_raw</span><span class="p">,</span>
        <span class="n">new_segs</span><span class="p">,</span> <span class="n">align_seq</span><span class="p">,</span> <span class="n">norm_signal</span><span class="p">,</span> <span class="n">scale_values</span><span class="p">,</span> <span class="n">corrected_group</span><span class="p">,</span>
        <span class="n">basecall_subgroup</span><span class="p">,</span> <span class="n">norm_type</span><span class="p">,</span> <span class="n">outlier_thresh</span><span class="p">,</span> <span class="n">compute_sd</span><span class="p">,</span>
        <span class="n">alignVals</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">align_info</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">old_segs</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">rna</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Write new fast5 group with re-squiggle data</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="c1"># compute event data before accessing fast5 file</span>
        <span class="k">if</span> <span class="n">compute_sd</span><span class="p">:</span>
            <span class="n">norm_means</span><span class="p">,</span> <span class="n">norm_stds</span> <span class="o">=</span> <span class="n">c_new_mean_stds</span><span class="p">(</span><span class="n">norm_signal</span><span class="p">,</span> <span class="n">new_segs</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">norm_means</span> <span class="o">=</span> <span class="n">c_new_means</span><span class="p">(</span><span class="n">norm_signal</span><span class="p">,</span> <span class="n">new_segs</span><span class="p">)</span>
            <span class="n">norm_stds</span> <span class="o">=</span> <span class="n">repeat</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">NAN</span><span class="p">)</span>

        <span class="c1"># had to shift to names formats numpy array specification due to</span>
        <span class="c1"># python2 numpy unicode issues. See discussion here:</span>
        <span class="c1"># https://github.com/numpy/numpy/issues/2407</span>
        <span class="n">event_data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
            <span class="nb">list</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">norm_means</span><span class="p">,</span> <span class="n">norm_stds</span><span class="p">,</span>
                     <span class="n">new_segs</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">np</span><span class="o">.</span><span class="n">diff</span><span class="p">(</span><span class="n">new_segs</span><span class="p">),</span> <span class="nb">list</span><span class="p">(</span><span class="n">align_seq</span><span class="p">))),</span>
            <span class="n">dtype</span><span class="o">=</span><span class="p">[(</span><span class="nb">str</span><span class="p">(</span><span class="s1">&#39;norm_mean&#39;</span><span class="p">),</span> <span class="s1">&#39;f8&#39;</span><span class="p">),</span> <span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="s1">&#39;norm_stdev&#39;</span><span class="p">),</span> <span class="s1">&#39;f8&#39;</span><span class="p">),</span>
                   <span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="s1">&#39;start&#39;</span><span class="p">),</span> <span class="s1">&#39;u4&#39;</span><span class="p">),</span> <span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="s1">&#39;length&#39;</span><span class="p">),</span> <span class="s1">&#39;u4&#39;</span><span class="p">),</span>
                   <span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="s1">&#39;base&#39;</span><span class="p">),</span> <span class="s1">&#39;S1&#39;</span><span class="p">)])</span>

        <span class="k">if</span> <span class="n">alignVals</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">r_align_vals</span><span class="p">,</span> <span class="n">g_align_vals</span> <span class="o">=</span> <span class="nb">zip</span><span class="p">(</span><span class="o">*</span><span class="n">alignVals</span><span class="p">)</span>
            <span class="n">np_read_align</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">chararray</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">alignVals</span><span class="p">))</span>
            <span class="n">np_read_align</span><span class="p">[:]</span> <span class="o">=</span> <span class="n">r_align_vals</span>
            <span class="n">np_genome_align</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">chararray</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">alignVals</span><span class="p">))</span>
            <span class="n">np_genome_align</span><span class="p">[:]</span> <span class="o">=</span> <span class="n">g_align_vals</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s1">&#39;Error computing new events&#39;</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">fast5_data</span><span class="p">,</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">):</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">fast5_data</span> <span class="o">=</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="n">fast5_data</span><span class="p">,</span> <span class="s1">&#39;r+&#39;</span><span class="p">)</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span>
                <span class="s1">&#39;Error opening file for new group writing. This should &#39;</span> <span class="o">+</span>
                <span class="s1">&#39;have been caught during the alignment phase. Check that &#39;</span> <span class="o">+</span>
                <span class="s1">&#39;there are no other tombo processes or processes &#39;</span> <span class="o">+</span>
                <span class="s1">&#39;accessing these HDF5 files running simultaneously.&#39;</span><span class="p">)</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="n">analysis_grp</span> <span class="o">=</span> <span class="n">fast5_data</span><span class="p">[</span><span class="s1">&#39;/Analyses&#39;</span><span class="p">]</span>
        <span class="n">corr_grp</span> <span class="o">=</span> <span class="n">analysis_grp</span><span class="p">[</span><span class="n">corrected_group</span><span class="p">]</span>
        <span class="c1"># add subgroup matching subgroup from original basecalls</span>
        <span class="n">corr_subgrp</span> <span class="o">=</span> <span class="n">corr_grp</span><span class="o">.</span><span class="n">create_group</span><span class="p">(</span><span class="n">basecall_subgroup</span><span class="p">)</span>
        <span class="n">corr_subgrp</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s1">&#39;status&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;success&#39;</span>
        <span class="n">corr_subgrp</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s1">&#39;rna&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">rna</span>
        <span class="n">corr_subgrp</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s1">&#39;shift&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">scale_values</span><span class="o">.</span><span class="n">shift</span>
        <span class="n">corr_subgrp</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s1">&#39;scale&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">scale_values</span><span class="o">.</span><span class="n">scale</span>
        <span class="n">corr_subgrp</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s1">&#39;lower_lim&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">scale_values</span><span class="o">.</span><span class="n">lower_lim</span>
        <span class="n">corr_subgrp</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s1">&#39;upper_lim&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">scale_values</span><span class="o">.</span><span class="n">upper_lim</span>
        <span class="n">corr_subgrp</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s1">&#39;norm_type&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">norm_type</span>
        <span class="n">corr_subgrp</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s1">&#39;outlier_threshold&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">outlier_thresh</span>

        <span class="c1"># store alignment statistics</span>
        <span class="n">corr_alignment</span> <span class="o">=</span> <span class="n">corr_subgrp</span><span class="o">.</span><span class="n">create_group</span><span class="p">(</span><span class="s1">&#39;Alignment&#39;</span><span class="p">)</span>
        <span class="n">corr_alignment</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s1">&#39;mapped_start&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">genome_location</span><span class="o">.</span><span class="n">Start</span>
        <span class="n">corr_alignment</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s1">&#39;mapped_end&#39;</span><span class="p">]</span> \
            <span class="o">=</span> <span class="n">genome_location</span><span class="o">.</span><span class="n">Start</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">new_segs</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span>
        <span class="n">corr_alignment</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s1">&#39;mapped_strand&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">genome_location</span><span class="o">.</span><span class="n">Strand</span>
        <span class="n">corr_alignment</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s1">&#39;mapped_chrom&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">genome_location</span><span class="o">.</span><span class="n">Chrom</span>

        <span class="k">if</span> <span class="n">align_info</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">corr_alignment</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s1">&#39;clipped_bases_start&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">align_info</span><span class="o">.</span><span class="n">ClipStart</span>
            <span class="n">corr_alignment</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s1">&#39;clipped_bases_end&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">align_info</span><span class="o">.</span><span class="n">ClipEnd</span>
            <span class="n">corr_alignment</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s1">&#39;num_insertions&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">align_info</span><span class="o">.</span><span class="n">Insertions</span>
            <span class="n">corr_alignment</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s1">&#39;num_deletions&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">align_info</span><span class="o">.</span><span class="n">Deletions</span>
            <span class="n">corr_alignment</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s1">&#39;num_matches&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">align_info</span><span class="o">.</span><span class="n">Matches</span>
            <span class="n">corr_alignment</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s1">&#39;num_mismatches&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">align_info</span><span class="o">.</span><span class="n">Mismatches</span>

        <span class="k">if</span> <span class="n">alignVals</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">corr_alignment</span><span class="o">.</span><span class="n">create_dataset</span><span class="p">(</span>
                <span class="s1">&#39;read_alignment&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">np_read_align</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="s2">&quot;gzip&quot;</span><span class="p">)</span>
            <span class="n">corr_alignment</span><span class="o">.</span><span class="n">create_dataset</span><span class="p">(</span>
                <span class="s1">&#39;genome_alignment&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">np_genome_align</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="s2">&quot;gzip&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">old_segs</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="c1"># store old segmentation in order to plot &quot;correction process&quot;</span>
            <span class="n">corr_alignment</span><span class="o">.</span><span class="n">create_dataset</span><span class="p">(</span>
                <span class="s1">&#39;read_segments&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">old_segs</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="s2">&quot;gzip&quot;</span><span class="p">)</span>

        <span class="c1"># Add Events to data frame with event means, SDs and lengths</span>
        <span class="n">corr_events</span> <span class="o">=</span> <span class="n">corr_subgrp</span><span class="o">.</span><span class="n">create_dataset</span><span class="p">(</span>
            <span class="s1">&#39;Events&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">event_data</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="s2">&quot;gzip&quot;</span><span class="p">)</span>
        <span class="n">corr_events</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span>
            <span class="s1">&#39;read_start_rel_to_raw&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">read_start_rel_to_raw</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span>
            <span class="s1">&#39;Error writing resquiggle information back into fast5 file.&#39;</span><span class="p">)</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="n">fast5_data</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span>
            <span class="s1">&#39;Error closing fast5 file after writing resquiggle information.&#39;</span><span class="p">)</span>
</div>
    <span class="k">return</span>


<span class="c1">###################################</span>
<span class="c1">###### Filter Main Functions ######</span>
<span class="c1">###################################</span>

<div class="viewcode-block" id="clear_filters_main"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.clear_filters_main">[docs]</a><span class="k">def</span> <span class="nf">clear_filters_main</span><span class="p">(</span><span class="n">args</span><span class="p">):</span>
    <span class="k">global</span> <span class="n">VERBOSE</span>
    <span class="n">VERBOSE</span> <span class="o">=</span> <span class="ow">not</span> <span class="n">args</span><span class="o">.</span><span class="n">quiet</span>

    <span class="k">for</span> <span class="n">fast5s_dir</span> <span class="ow">in</span> <span class="n">args</span><span class="o">.</span><span class="n">fast5_basedirs</span><span class="p">:</span>
        <span class="n">clear_filters</span><span class="p">(</span><span class="n">fast5s_dir</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">corrected_group</span><span class="p">)</span>
</div>
    <span class="k">return</span>

<div class="viewcode-block" id="filter_stuck_main"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.filter_stuck_main">[docs]</a><span class="k">def</span> <span class="nf">filter_stuck_main</span><span class="p">(</span><span class="n">args</span><span class="p">):</span>
    <span class="k">global</span> <span class="n">VERBOSE</span>
    <span class="n">VERBOSE</span> <span class="o">=</span> <span class="ow">not</span> <span class="n">args</span><span class="o">.</span><span class="n">quiet</span>

    <span class="n">obs_filter</span> <span class="o">=</span> <span class="n">parse_obs_filter</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">obs_per_base_filter</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">fast5s_dir</span> <span class="ow">in</span> <span class="n">args</span><span class="o">.</span><span class="n">fast5_basedirs</span><span class="p">:</span>
        <span class="n">filter_reads</span><span class="p">(</span><span class="n">fast5s_dir</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">corrected_group</span><span class="p">,</span> <span class="n">obs_filter</span><span class="p">)</span>
</div>
    <span class="k">return</span>

<div class="viewcode-block" id="filter_coverage_main"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.filter_coverage_main">[docs]</a><span class="k">def</span> <span class="nf">filter_coverage_main</span><span class="p">(</span><span class="n">args</span><span class="p">):</span>
    <span class="k">global</span> <span class="n">VERBOSE</span>
    <span class="n">VERBOSE</span> <span class="o">=</span> <span class="ow">not</span> <span class="n">args</span><span class="o">.</span><span class="n">quiet</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="mi">0</span> <span class="o">&lt;</span> <span class="n">args</span><span class="o">.</span><span class="n">percent_to_filter</span> <span class="o">&lt;</span> <span class="mi">100</span><span class="p">:</span>
        <span class="n">_error_message_and_exit</span><span class="p">(</span>
            <span class="s1">&#39;--percent-to-filter must be between 0 and 100.&#39;</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">fast5s_dir</span> <span class="ow">in</span> <span class="n">args</span><span class="o">.</span><span class="n">fast5_basedirs</span><span class="p">:</span>
        <span class="n">filter_reads_for_coverage</span><span class="p">(</span>
            <span class="n">fast5s_dir</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">corrected_group</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">percent_to_filter</span> <span class="o">/</span> <span class="mf">100.0</span><span class="p">)</span>
</div>
    <span class="k">return</span>


<span class="c1">##################################</span>
<span class="c1">###### Annotate FAST5s Main ######</span>
<span class="c1">##################################</span>

<div class="viewcode-block" id="annotate_reads_with_fastq_main"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.annotate_reads_with_fastq_main">[docs]</a><span class="k">def</span> <span class="nf">annotate_reads_with_fastq_main</span><span class="p">(</span><span class="n">args</span><span class="p">):</span>
    <span class="k">global</span> <span class="n">VERBOSE</span>
    <span class="n">VERBOSE</span> <span class="o">=</span> <span class="ow">not</span> <span class="n">args</span><span class="o">.</span><span class="n">quiet</span>

    <span class="n">fast5s_basedir</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">args</span><span class="o">.</span><span class="n">fast5_basedir</span> <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">fast5_basedir</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)</span> <span class="k">else</span>
        <span class="n">args</span><span class="o">.</span><span class="n">fast5_basedir</span> <span class="o">+</span> <span class="s1">&#39;/&#39;</span><span class="p">)</span>
    <span class="n">fast5s_read_ids</span> <span class="o">=</span> <span class="n">get_read_ids_and_prep_fastq_slot</span><span class="p">(</span>
        <span class="n">fast5s_basedir</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">basecall_group</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">basecall_subgroup</span><span class="p">,</span>
        <span class="n">args</span><span class="o">.</span><span class="n">overwrite</span><span class="p">)</span>
    <span class="n">fastq_slot</span> <span class="o">=</span> <span class="s1">&#39;/&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">((</span><span class="s1">&#39;/Analyses&#39;</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">basecall_group</span><span class="p">,</span>
                           <span class="n">args</span><span class="o">.</span><span class="n">basecall_subgroup</span><span class="p">))</span>
    <span class="n">annotate_with_fastqs</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">fastq_filenames</span><span class="p">,</span> <span class="n">fast5s_read_ids</span><span class="p">,</span> <span class="n">fastq_slot</span><span class="p">)</span>
</div>
    <span class="k">return</span>


<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s1">&#39;__main__&#39;</span><span class="p">:</span>
    <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span>
        <span class="s1">&#39;This is a module. See commands with `tombo -h`&#39;</span><span class="p">)</span>
</pre></div>

           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, Oxford Nanopore Technologies.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../',
            VERSION:'1.2',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="../../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>